{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "import pandas as pd\n",
    "import glob\n",
    "import os\n",
    "import os.path as op\n",
    "\n",
    "import networkx as nx\n",
    "import random\n",
    "import community\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def prep_sag_fastani(fastani_out):\n",
    "    '''\n",
    "    Parses sag ids, \n",
    "    removes self-hits, \n",
    "    keeps only one pair-wise comparison (the one with the highest ANI) per SAG pair\n",
    "    '''\n",
    "    \n",
    "    df = pd.read_csv(fastani_out, sep=\"\\t\", names=['c1','c2','ani','frag_mappings','total_frags'])\n",
    "\n",
    "    df['genome1'] = [op.basename(i).split(\"_\")[0] for i in df['c1']]\n",
    "    df['genome2'] = [op.basename(i).split(\"_\")[0] for i in df['c2']]\n",
    "\n",
    "    df = df[df['genome1'] != df['genome2']].copy()\n",
    "\n",
    "    pairs = []\n",
    "\n",
    "    for i, l in df.iterrows():\n",
    "        genome_pair_string = \"_\".join(sorted([l['genome1'], l['genome2']]))\n",
    "        pairs.append(genome_pair_string)\n",
    "\n",
    "    df['pair'] = pairs\n",
    "\n",
    "    dedupdf = df.sort_values(by='ani', ascending=False).drop_duplicates(subset='pair', keep='first')\n",
    "    return dedupdf\n",
    "\n",
    "def cluster_fastani(df, threshold = 90):\n",
    "    '''\n",
    "    clusters SAGs over specified ani threshold\n",
    "    '''\n",
    "    subdf = df[df['ani'] >= threshold]\n",
    "    nodes = list(set(list(subdf['genome1']) + list(subdf['genome2'])))\n",
    "    g = nx.Graph()\n",
    "    g.add_nodes_from(nodes)\n",
    "    edges = [(l.genome1, l.genome2) for i, l in subdf.iterrows()]\n",
    "    g.add_edges_from(edges)\n",
    "    partition = community.best_partition(g)\n",
    "    clusters = pd.DataFrame.from_dict(partition, orient='index').reset_index().rename(columns={\"index\":'sample', 0:'group'})\n",
    "    return clusters\n",
    "\n",
    "def networkx_graph(df, threshold = 90):\n",
    "    '''\n",
    "    Creates networkx graph from fastani \n",
    "    '''\n",
    "    subdf = df[df['ani'] >= threshold]\n",
    "    nodes = list(set(list(subdf['genome1']) + list(subdf['genome2'])))\n",
    "    g = nx.Graph()\n",
    "    g.add_nodes_from(nodes)\n",
    "    edges = [(l.genome1, l.genome2) for i, l in subdf.iterrows()]\n",
    "    g.add_edges_from(edges)\n",
    "    return g\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fastANI output file. The path to this file needs to be added\n",
    "fastani_out = ''\n",
    "\n",
    "# File that has metadata related to each SAG. The columns below need to be present\n",
    "sag_info_file = ''\n",
    "\n",
    "#loading SAG table, pulling out relevant columns\n",
    "\n",
    "# THE SSU_classification_1 here and anywhere else that it occurs probably needs to be changed to GTDB_classification or whatever that column title is\n",
    "# I believe that there are five occurances of this that need to be changed\n",
    "infodf = pd.read_csv(sag_info_file)[['name','Plate','SSU_classification_1','well_type']]\n",
    "\n",
    "#Cleanup in case there are any 'no drop' or '10 cell' wells in this file\n",
    "# i noticed that some control SAGs were included in the fastani analysis, so pulling a list of those out for removal from the fastani output\n",
    "control_sags = list(infodf[infodf['well_type'] != '1 cell']['name'])\n",
    "\n",
    "# just checking\n",
    "print(\"there are\", len(control_sags), \"control SAGs that we don't want to be included in the fastani analysis\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "206441"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load fastani output\n",
    "df = prep_sag_fastani(fastani_out)\n",
    "df.to_csv(\"./tables/gom_fastani_organized.csv\", index=False)\n",
    "print('there are ', len(df),' genome pairs to be included in the analysis')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AH-135-L02',\n",
       " 'AH-224-M21',\n",
       " 'AH-229-E23',\n",
       " 'AH-229-J06',\n",
       " 'AH-229-J20',\n",
       " 'AH-229-K20',\n",
       " 'AH-269-A19',\n",
       " 'AH-269-D11',\n",
       " 'AH-269-E18',\n",
       " 'AH-269-J11',\n",
       " 'AH-269-J18',\n",
       " 'AH-269-N03',\n",
       " 'AH-545-J02',\n",
       " 'AH-545-M06',\n",
       " 'AH-578-G21',\n",
       " 'AH-601-O09',\n",
       " 'AH-603-G11',\n",
       " 'AH-603-J06',\n",
       " 'AH-654-B05',\n",
       " 'AH-654-G08',\n",
       " 'AH-654-G16',\n",
       " 'AH-654-I06',\n",
       " 'AH-657-P13',\n",
       " 'AH-704-A15',\n",
       " 'AH-704-D13',\n",
       " 'AH-707-D22',\n",
       " 'AH-707-K02',\n",
       " 'AH-707-P01',\n",
       " 'AH-709-P09',\n",
       " 'AH-770-O06',\n",
       " 'AH-771-F11',\n",
       " 'AH-773-A02',\n",
       " 'AH-773-M21',\n",
       " 'AH-775-I05',\n",
       " 'AH-779-K06',\n",
       " 'AH-781-M20',\n",
       " 'AH-135-G04',\n",
       " 'AH-229-B13',\n",
       " 'AH-700-P17',\n",
       " 'AH-704-A23',\n",
       " 'AH-704-D19',\n",
       " 'AH-704-F09',\n",
       " 'AH-704-O20',\n",
       " 'AH-707-N16',\n",
       " 'AH-224-O06',\n",
       " 'AH-551-O09',\n",
       " 'AH-657-F10',\n",
       " 'AH-541-B10',\n",
       " 'AH-707-M15',\n",
       " 'AH-709-A09',\n",
       " 'AH-709-B15',\n",
       " 'AH-709-J11',\n",
       " 'AH-141-I11',\n",
       " 'AH-883-E03',\n",
       " 'AH-883-F02',\n",
       " 'AH-888-C02',\n",
       " 'AH-787-N10',\n",
       " 'AH-305-P10',\n",
       " 'AD-866-D03',\n",
       " 'AD-866-J03',\n",
       " 'AD-866-I18',\n",
       " 'AD-866-D18',\n",
       " 'AD-867-P11',\n",
       " 'AD-867-N03',\n",
       " 'AD-867-N07',\n",
       " 'AD-867-P02',\n",
       " 'AAA160-B08',\n",
       " 'AH-704-I07',\n",
       " 'AD-867-K16',\n",
       " 'AD-867-B05',\n",
       " 'AD-867-D18',\n",
       " 'AD-866-L13',\n",
       " 'AH-736-J23',\n",
       " 'AH-736-P06']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### THIS CELL CAN BE REMOVED IF QC ALREADY DONE\n",
    "\n",
    "#This is a folder of SAGs that I had initially identified as contaminated and wanted removed from the clustering\n",
    "\n",
    "# generating a list of contaminant/cosorts to remove from fastani output \n",
    "bad_sags = [op.basename(i).split(\"_\")[0] for i in glob.glob('/mnt/scgc/simon/microg2p/Data/SAG_contigs/contaminated_SAG_contigs/*_contigs.fasta')]\n",
    "\n",
    "bad_sags = []\n",
    "\n",
    "#Adding additional BAD SAGs that were identified at a later point\n",
    "with open(\"/mnt/scgc/simon/microg2p/Data/plate_lists/contaminated_SAGs.txt\") as ih:\n",
    "    \n",
    "    for l in ih:\n",
    "        bad_sags.append(l.strip())\n",
    "        \n",
    "bad_sags\n",
    "\n",
    "# just checking\n",
    "print(\"There are\", len(bad_sags), \"contaminated SAGs to exclude from analysis\")\n",
    "print(bad_sags[:3])\n",
    "\n",
    "\n",
    "# remove rows that include contaminated sags and control SAGs\n",
    "remove_sags = bad_sags + control_sags\n",
    "\n",
    "df = df[~df['genome1'].isin(remove_sags) & ~df['genome2'].isin(remove_sags)]\n",
    "\n",
    "print('there are', len(df), 'genome pairs that are included in the analysis after the removal of bad SAGs')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0, 0.5, '# Genome Pairs')"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZIAAAEGCAYAAABPdROvAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAbvklEQVR4nO3dcbRdZX3m8e8jUQRtFEjAmAQTIWhDljNKClFnEMUZqGMJzoJpgpbUpk3LQhHtVInOKrpW00K1atGBtaKhBFQwRSvpLEExVqkdIF5UhICR1GC4kJLLQClqDSQ888d+rxxuzj053H3POZx7ns9ad529f3vvs9+XTe7vvvvd+31lm4iIiIl6Tq8LEBER/S2JJCIiakkiiYiIWpJIIiKiliSSiIioZVqvC9BtM2bM8Lx583pdjIiIvnLbbbc9ZHtms20Dl0jmzZvH0NBQr4sREdFXJP10vG25tRUREbUkkURERC1JJBERUUsSSURE1JJEEhERtSSRRERELUkkERFRSxJJRETUkkQSERG1DNyb7TG+L9y6Y9xtZ51wZBdLEhH9JC2SiIioJYkkIiJqSSKJiIhaOpZIJF0uaZekO8fE3y1pq6Qtkv6yIb5a0ray7ZSG+HGS7ijbLpGkEj9Q0hdL/FZJ8zpVl4iIGF8nO9uvAD4NXDkakPRGYCnwKtu7JR1e4guBZcCxwEuBb0g6xvZe4DJgFXAL8FXgVOB6YCXwiO2jJS0DLgZ+u4P1mRJadahHRExEx1oktm8CHh4TPge4yPbuss+uEl8KXGN7t+3twDbgeEmzgOm2b7ZtqqR0esMx68vytcDJo62ViIjonm73kRwD/OdyK+rbkn6jxGcD9zXsN1xis8vy2PjTjrG9B3gUOKzZSSWtkjQkaWhkZGTSKhMREd1PJNOAQ4AlwJ8AG0orollLwi3i7Gfb04P2WtuLbS+eObPpTJERETFB3U4kw8CXXdkMPAnMKPG5DfvNAR4o8TlN4jQeI2ka8CL2vZUWEREd1u1E8hXgTQCSjgGeBzwEbASWlSex5gMLgM22dwKPSVpSWi5nA9eV79oIrCjLZwDfLP0oERHRRR17akvS1cBJwAxJw8CFwOXA5eWR4MeBFeWX/xZJG4C7gD3AueWJLag66K8ADqJ6Wuv6El8HXCVpG1VLZFmn6hIREePrWCKxvXycTe8YZ/81wJom8SFgUZP4L4Ez65QxIiLqy5vtERFRSxJJRETUkkQSERG1JJFEREQtSSQREVFLEklERNSSRBIREbUkkURERC1JJBERUUsSSURE1JJEEhERtSSRRERELUkkERFRSxJJRETUkkQSERG1JJFEREQtHUskki6XtKvMhjh22/+UZEkzGmKrJW2TtFXSKQ3x4yTdUbZdUqbcpUzL+8USv1XSvE7VJSIixtfJFskVwKljg5LmAv8F2NEQW0g1Ve6x5ZhLJR1QNl8GrKKax31Bw3euBB6xfTTwCeDijtQiIiJa6lgisX0T1VzqY30CeD/ghthS4Brbu21vB7YBx0uaBUy3fXOZ2/1K4PSGY9aX5WuBk0dbKxER0T1d7SORdBpwv+3bx2yaDdzXsD5cYrPL8tj4046xvQd4FDhsnPOukjQkaWhkZKR2PSIi4ildSySSDgY+BPxps81NYm4Rb3XMvkF7re3FthfPnDmzneJGRESbutkiOQqYD9wu6V5gDvA9SS+hamnMbdh3DvBAic9pEqfxGEnTgBfR/FZaRER0UNcSie07bB9ue57teVSJ4DW2/wXYCCwrT2LNp+pU32x7J/CYpCWl/+Ns4LrylRuBFWX5DOCbpR8lIiK6qJOP/14N3Ay8QtKwpJXj7Wt7C7ABuAu4ATjX9t6y+Rzgs1Qd8P8MXF/i64DDJG0D3gdc0JGKRERES9M69cW2l+9n+7wx62uANU32GwIWNYn/EjizXikjIqKuvNkeERG1JJFEREQtSSQREVFLEklERNSSRBIREbUkkURERC1JJBERUUsSSURE1NKxFxIjImLivnDrjnG3nXXCkV0syf6lRRIREbUkkURERC1JJBERUUsSSURE1JJEEhERtSSRRERELUkkERFRSydnSLxc0i5JdzbEPirpR5J+KOnvJL24YdtqSdskbZV0SkP8OEl3lG2XlCl3KdPyfrHEb5U0r1N1iYiI8XWyRXIFcOqY2I3AItuvAn4MrAaQtBBYBhxbjrlU0gHlmMuAVVTzuC9o+M6VwCO2jwY+AVzcsZpERMS4OpZIbN8EPDwm9nXbe8rqLcCcsrwUuMb2btvbqeZnP17SLGC67ZttG7gSOL3hmPVl+Vrg5NHWSkREdE8v+0h+D7i+LM8G7mvYNlxis8vy2PjTjinJ6VHgsGYnkrRK0pCkoZGRkUmrQERE9CiRSPoQsAf4/GioyW5uEW91zL5Be63txbYXz5w585kWNyIiWthvIpF0lKQDy/JJks5r7CR/piStAN4KvL3croKqpTG3Ybc5wAMlPqdJ/GnHSJoGvIgxt9IiIqLz2mmRfAnYK+loYB0wH/jCRE4m6VTgA8Bptn/RsGkjsKw8iTWfqlN9s+2dwGOSlpT+j7OB6xqOWVGWzwC+2ZCYIiKiS9oZRv5J23skvQ34pO1PSfr+/g6SdDVwEjBD0jBwIdVTWgcCN5Z+8Vts/5HtLZI2AHdR3fI61/be8lXnUD0BdhBVn8pov8o64CpJ26haIsvaqXBEREyudhLJE5KWU/31/1sl9tz9HWR7eZPwuhb7rwHWNIkPAYuaxH8JnLm/ckRERGe1c2vrncBrgTW2t5dbT5/rbLEiIqJftGyRlJcCP2j7HaOx8p7HRZ0uWERE9IeWLZLSTzFT0vO6VJ6IiOgz7fSR3Av8k6SNwM9Hg7Y/3qlCRURE/2gnkTxQfp4D/FpnixMREf1mv4nE9ke6UZCIiOhP4yYSSZ+0fb6kv6fJ0CO2T+toySIioi+0apFcVT4/1o2CREREfxo3kdi+rXx+u3vFiYiIfrPfPhJJC4C/ABYCzx+N2355B8sVERF9op032/+GapbCPcAbqSaXuqrlERERMTDaSSQH2d4EyPZPbX8YeFNnixUREf2infdIfinpOcA9kt4F3A8c3tliRUREv2inRXI+cDBwHnAc8A6emgckIiIG3P4GbZxJ9Q7JNNvDVCMBR0RE/Mq4LRJJvw9sAT4F/EhSXkCMiIh9tLq1dT5wrO3XAq+jmt2wbZIul7RL0p0NsUMl3SjpnvJ5SMO21ZK2Sdoq6ZSG+HGS7ijbLilT7lKm5f1iid8qad4zKV9EREyOVonkcdsjALZ/QjVF7jNxBXDqmNgFwCbbC4BNZR1JC6mmyj22HHNpmQsFqkePV1HN476g4TtXAo/YPhr4BHDxMyxfRERMglZ9JHMkXTLeuu3zWn2x7ZuatBKWUs3jDrAe+BbwgRK/xvZuYHuZh/14SfcC023fDCDpSuB0qnnblwIfLt91LfBpSbK9z7hgERHROa0SyZ+MWb9tEs53hO2dALZ3Shp9jHg2cEvDfsMl9kRZHhsfPea+8l17JD0KHAY8NPakklZRtWo48sgjJ6EaERExqtVYW+u7WA41K0KLeKtj9g3aa4G1AIsXL06LJSJiErXzQuJkelDSrNIamQXsKvFhYG7DfnOoJtMaLstj443HDEuaBrwIeLiThY+ImExfuHVHr4swKdp5IXEybeSplxlXANc1xJeVJ7HmU3Wqby63wR6TtKQ8rXX2mGNGv+sM4JvpH4mI6L6OtUgkXU3VsT5D0jBwIXARsEHSSmAHcCaA7S2SNgB3UQ0Oea7tveWrzqF6Auwgqk7260t8HXBV6Zh/mOqpr4iIKa9VS+asE7rfD9zOMPLHUD2Ce4TtRZJeBZxm+89aHWd7+TibTh5n/zXAmibxIWBRk/gvKYkoIiJ6p51bW5+hehnxCQDbPyR//UdERNFOIjnY9uYxsT2dKExERPSfdhLJQ5KOojxaK+kMYGdHSxUREX2jnc72c6newXilpPuB7VRDyUdEROw/kZRxtt4s6QXAc2w/1vliRUREv2jnqa0XU72/MQ+YVgbf3e9YWxERMRjaubX1VapxsO4AnuxscSIiot+0k0ieb/t9HS9JRET0pXae2rpK0h9ImlUmpjpU0qEdL1lERPSFdlokjwMfBT7EU6PrGnh5pwoVEdFPpsrgixPVTiJ5H3C07X3m+YiIGBSDnixaaefW1hbgF50uSERE9Kd2WiR7gR9I+gdg92gwj/9GRAS0l0i+Un4iIiL20c6b7eslPQ84poS22n6is8WKiIh+0c6b7ScB64F7qeZJnytphe2bOlu0iIjoB+10tv8V8F9tv8H2icApwCfqnFTSeyVtkXSnpKslPb+8n3KjpHvK5yEN+6+WtE3SVkmnNMSPk3RH2XaJRsdviYiIrmknkTzX9tbRFds/Bp470RNKmg2cByy2vQg4gGqirAuATbYXAJvKOpIWlu3HAqcCl0o6oHzdZcAqqjneF5TtERHRRe0kkiFJ6ySdVH4+A9xW87zTgIMkTQMOBh4AllLdQqN8nl6WlwLX2N5tezuwDThe0ixguu2bbRu4suGYiIjoknYSyTlU75KcB7wHuAv4o4me0Pb9wMeAHVQTZD1q++tUc8LvLPvsBA4vh8wG7mv4iuESm12Wx8b3IWmVpCFJQyMjIxMtekRENNHOU1u7gY+Xn9pK38dSYD7wr8DfSmo1UVazfg+3iO8btNdSTc7F4sWLm+4TERET085TW68HPgy8rHF/2xMda+vNwHbbI+X7vwy8DnhQ0izbO8ttq11l/2FgbsPxc6huhQ2X5bHxiIjoonZeSFwHvJeqX2TvJJxzB7BE0sHAvwMnA0PAz4EVwEXl87qy/0bgC5I+DryUqlN9s+29kh6TtAS4lWryrU9NQvmiifHGGTrrhCO7XJKIeLZpJ5E8avv6yTqh7VslXQt8D9gDfJ/qttMLgQ2SVlIlmzPL/lskbaDqm9kDnGt7NKGdA1wBHARcX34iIqKL2kkk/yDpo8CXefpYW9+b6EltXwhcOCa8m6p10mz/NcCaJvEhYNFEyxEREfW1k0hOKJ+LG2IG3jT5xYmIiH7TzlNbb+xGQSIioj+189TWEcCfAy+1/ZvlTfPX2l7X8dLFhGQCnojopnZeSLwC+BrVE1MAPwbO71SBIiKiv7STSGbY3gA8CWB7D5PzGHBEREwB7SSSn0s6jPLWeHlv49GOlioiIvpGO09tvY/qpcCjJP0TMBM4o6OlioiIvtHOU1vfk/QG4BVU41tlhsSIiPiVcW9tSZouaQH8ql9kIfAaYHl5kisiIqJlH8nHgNc3rP8F1UuJJwIf6WShIiKif7S6tfUbwB82rD9m+zwASd/paKkiIqJvtGqRTCszD476nYblF3eoPBER0WdaJZInJb1kdMX2nfCrOdef7HTBIiKiP7RKJB8F/l7SiZJ+rfy8AfhK2RYRETF+H4ntz0l6CPgz4FiqFxK3AH86mfOTREREf2v5HontG4AbulSWiIieyoCnE9POECmTTtKLJV0r6UeS7pb0WkmHSrpR0j3l85CG/VdL2iZpq6RTGuLHSbqjbLtEknpRn4iIQdaTRAL8NXCD7VcC/wG4G7gA2GR7AbCprFOGrV9GdXvtVOBSSQeU77kMWEU1j/uCsj0iIrqo64lE0nSqlxrXAdh+3Pa/AkuB9WW39cDpZXkpcI3t3ba3A9uA4yXNAqbbvrk8pnxlwzEREdEl+00kkv5Xw/KBk3DOlwMjwN9I+r6kz0p6AXCE7Z0A5fPwsv9s4L6G44dLbHZZHhtvVodVkoYkDY2MjExCFSIiYlSrsbbeL+m1PH2k35sn4ZzTqMbsusz2q4GfU25jjVeUJjG3iO8btNfaXmx78cyZM59peSMiooVWLZKtwJnAyyX9o6S1wGGSXlHznMPAsO1by/q1VInlwXK7ivK5q2H/uQ3HzwEeKPE5TeIREdFFrRLJI8AHqfokTgIuKfELJP3fiZ7Q9r8A9zUkpJOBu6jmPFlRYiuA68ryRmCZpAMlzafqVN9cbn89JmlJeVrr7IZjIiKiS1q9R3IqcCFwFPBx4Hbg57bfOQnnfTfweUnPA34CvJMqqW2QtBLYQdUawvYWSRuoks0e4Fzbo1P9nkM1p/xBwPXlJyIiuqjVm+0fBJB0O/A54NXAzDLy7yO2f2uiJ7X9A6oh6cc6eZz91wBrmsSHgEUTLUdERNTXzlS7X7P9XeC7ks6x/Z8kzeh0waK1vIEbEc8W+3381/b7G1Z/t8Qe6lSBIiKivzyjFxJt396pgkRERH/q1RApERExRSSRRERELUkkERFRSxJJRETUkkQSERG1JJFEREQtSSQREVFLO2+2R4yr1Rv2Z51wZBdLEhG9khZJRETUkkQSERG1JJFEREQtSSQREVFLOtsjBlQelIjJ0rMWiaQDJH1f0v8p64dKulHSPeXzkIZ9V0vaJmmrpFMa4sdJuqNsu6RMuRsREV3UyxbJe4C7gell/QJgk+2LJF1Q1j8gaSGwDDgWeCnwDUnHlOl2LwNWAbcAX6WaHnjKTLebyasioh/0pEUiaQ7w34DPNoSXAuvL8nrg9Ib4NbZ3294ObAOOlzQLmG77ZtsGrmw4JiIiuqRXt7Y+CbwfeLIhdoTtnQDl8/ASnw3c17DfcInNLstj4/uQtErSkKShkZGRyalBREQAPbi1JemtwC7bt0k6qZ1DmsTcIr5v0F4LrAVYvHhx030ipqLcHo1u6EUfyeuB0yS9BXg+MF3S54AHJc2yvbPcttpV9h8G5jYcPwd4oMTnNIlHRE1T+YmuJNfJ1/VbW7ZX255jex5VJ/o3bb8D2AisKLutAK4ryxuBZZIOlDQfWABsLre/HpO0pDytdXbDMRER0SXPpvdILgI2SFoJ7ADOBLC9RdIG4C5gD3BueWIL4BzgCuAgqqe1pswTWxER/aKnicT2t4BvleX/B5w8zn5rgDVN4kPAos6VMCIi9idDpERERC1JJBERUUsSSURE1JJEEhERtTybntqKiD4wld8xiYlJiyQiImpJiySmjPylHNEbaZFEREQtaZFEtJBWTsT+JZH02FQeQC6/hCMGQxJJ9JWpnHgj+lUSSURMOfmDo7uSSGLg5ZdO/8q1e3ZIIomYoPQBRVSSSGIgTOW/XKdy3aI/JJFEdMBEfrlPhVZMWmmDqeuJRNJc4ErgJcCTwFrbfy3pUOCLwDzgXuB/2H6kHLMaWAnsBc6z/bUSP46nZkj8KvAe2+5mfSKis9LievbrxZvte4A/tv3rwBLgXEkLgQuATbYXAJvKOmXbMuBY4FTgUkkHlO+6DFhFNY/7grI9IiK6qOstEts7gZ1l+TFJdwOzgaXASWW39VRT8H6gxK+xvRvYLmkbcLyke4Hptm8GkHQlcDqZtz0iBlgvbi/2tI9E0jzg1cCtwBElyWB7p6TDy26zgVsaDhsusSfK8th4s/Osomq5cOSRuU8bz07pX4h+1bNBGyW9EPgScL7tf2u1a5OYW8T3DdprbS+2vXjmzJnPvLARETGunrRIJD2XKol83vaXS/hBSbNKa2QWsKvEh4G5DYfPAR4o8TlN4tEH0oEaMXV0vUUiScA64G7bH2/YtBFYUZZXANc1xJdJOlDSfKpO9c3lNthjkpaU7zy74ZiIiOiSXrRIXg/8DnCHpB+U2AeBi4ANklYCO4AzAWxvkbQBuIvqia9zbe8tx53DU4//Xk862mOKmgotuPQBTV29eGrrOzTv3wA4eZxj1gBrmsSHgEWTV7qI6IWpkCgHWWZIjIiIWpJIIiKiloy11QVptkfEVJYWSURE1JJEEhERtSSRRERELUkkERFRSxJJRETUkkQSERG1JJFEREQtSSQREVFLEklERNSSRBIREbUkkURERC1JJBERUUsGbZwkGZgxIgZV37dIJJ0qaaukbZIu6HV5IiIGTV8nEkkHAP8b+E1gIbBc0sLelioiYrD0dSIBjge22f6J7ceBa4ClPS5TRMRA6fc+ktnAfQ3rw8AJY3eStApYVVZ/JmnrON83A3hoUkvYPwa17oNab0jdB67ub69X75eNt6HfE4maxLxPwF4LrN3vl0lDthdPRsH6zaDWfVDrDan7INa9U/Xu91tbw8DchvU5wAM9KktExEDq90TyXWCBpPmSngcsAzb2uEwREQOlr29t2d4j6V3A14ADgMttb6nxlfu9/TWFDWrdB7XekLoPoo7UW/Y+XQoRERFt6/dbWxER0WNJJBERUcvAJhJJr5D0g4aff5N0vqQPS7q/If6WXpd1skl6r6Qtku6UdLWk50s6VNKNku4pn4f0upydME7dB+Gav6fUeYuk80tsUK55s7pPyWsu6XJJuyTd2RAb9zpLWl2Gl9oq6ZQJnzd9JL8aauV+qpcZ3wn8zPbHeluqzpA0G/gOsND2v0vaAHyVaoiZh21fVMYsO8T2B3pZ1snWou7zmNrXfBHVqA/HA48DNwDnAH/A1L/m49X97UzBay7pROBnwJW2F5XYX9LkOpfhpK6m+m/zUuAbwDG29z7T8w5si2SMk4F/tv3TXhekS6YBB0maBhxM9e7NUmB92b4eOL1HZeu0ZnWf6n4duMX2L2zvAb4NvI3BuObj1X1Ksn0T8PCY8HjXeSlwje3dtrcD26iSyjOWRFJZRpWZR71L0g9LM3FKNfdt3w98DNgB7AQetf114AjbO8s+O4HDe1fKzmhRd5jC1xy4EzhR0mGSDgbeQvUi75S/5oxfd5ja17zReNe52RBTsydygoFPJOVFxtOAvy2hy4CjgP9I9cvmr3pUtI4o/2CWAvOpmrMvkPSO3paqO1rUfUpfc9t3AxcDN1Ld2rkd2NPTQnVJi7pP6WvepraGmGrHwCcSqiHov2f7QQDbD9rea/tJ4DNMsKn3LPZmYLvtEdtPAF8GXgc8KGkWQPnc1cMydkrTug/ANcf2OtuvsX0i1a2PexiMa9607oNwzRuMd50nbYipJBJYTsNtrdH/4MXbqJrGU8kOYImkgyWJqn/obqqhZVaUfVYA1/WofJ3UtO4DcM2RdHj5PBL471T/zw/CNW9a90G45g3Gu84bgWWSDpQ0H1gAbJ7ICQb6qa1yz/Q+4OW2Hy2xq6iauwbuBf5w9P7iVCHpI8BvUzXxvw/8PvBCYANwJNUv3DNtj+2063vj1P2zTP1r/o/AYcATwPtsb5J0GINxzZvVfUr+O5d0NXAS1XDxDwIXAl9hnOss6UPA71H9ezjf9vUTOu8gJ5KIiKgvt7YiIqKWJJKIiKgliSQiImpJIomIiFqSSCIiopYkkoguk/Q2SZb0yrI+r6y/u2GfT0v63bJ8haQzelTciP1KIonovuVUoxAva4jtAt5ThuyJ6CtJJBFdJOmFwOuBlTw9kYwAm3jqDeSIvpFEEtFdpwM32P4x8LCk1zRsuwj44zI/TkTfSCKJ6K7lVBMtUT6Xj24oc0JsBs7qQbkiJmxarwsQMSjK2FZvAhZJMnAA1VhPlzbs9ufAtcBN3S9hxMSkRRLRPWdQTYH6MtvzbM8FtlMN3w2A7R8BdwFv7VEZI56xJJKI7lkO/N2Y2JeAD46JraEhuUQ822X034iIqCUtkoiIqCWJJCIiakkiiYiIWpJIIiKiliSSiIioJYkkIiJqSSKJiIha/j9Zzwnc0AUZ9QAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# just checking\n",
    "\n",
    "sns.distplot(df['ani'], kde=False)\n",
    "plt.xlabel('ANI')\n",
    "plt.ylabel('# Genome Pairs')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Eyeballing ANI thresholds to look into... will use all values (aka threshold=0), 83, 87 and 93."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>group87</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>AH-700-O11</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>AH-709-G13</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>AH-578-K18</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>AH-578-M09</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>AH-700-A04</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3939</td>\n",
       "      <td>AH-269-I11</td>\n",
       "      <td>176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3940</td>\n",
       "      <td>AH-229-D18</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3941</td>\n",
       "      <td>AH-141-N22</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3942</td>\n",
       "      <td>AH-578-K15</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3943</td>\n",
       "      <td>AH-578-M11</td>\n",
       "      <td>73</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3944 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          sample  group87\n",
       "0     AH-700-O11        0\n",
       "1     AH-709-G13        1\n",
       "2     AH-578-K18        2\n",
       "3     AH-578-M09        3\n",
       "4     AH-700-A04        4\n",
       "...          ...      ...\n",
       "3939  AH-269-I11      176\n",
       "3940  AH-229-D18       24\n",
       "3941  AH-141-N22        9\n",
       "3942  AH-578-K15       28\n",
       "3943  AH-578-M11       73\n",
       "\n",
       "[3944 rows x 2 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Main ANI threshoild it will produce the output files at this level modify if necessary \n",
    "threshold = 87\n",
    "\n",
    "clusters = cluster_fastani(df, threshold = threshold)\n",
    "clusters.rename(columns = {'group':'group{}'.format(threshold)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>AH-700-O11</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>AH-709-G13</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>AH-578-K18</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>AH-578-M09</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>AH-700-A04</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       sample  group\n",
       "0  AH-700-O11      0\n",
       "1  AH-709-G13      1\n",
       "2  AH-578-K18      2\n",
       "3  AH-578-M09      3\n",
       "4  AH-700-A04      4"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# just checking \n",
    "\n",
    "clusters.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# merging the cluster assignments with phylogenetic information\n",
    "\n",
    "cdf = clusters.merge(infodf, left_on='sample', right_on='name', how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group</th>\n",
       "      <th>SSU_classification_1</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>k__Eukaryota_(Chloroplast);p__?;c__?;o__?;f__?...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>no SSU regions found</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>no SSUs were classified</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>k__Bacteria;p__Proteobacteria;c__Alphaproteoba...</td>\n",
       "      <td>160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>no SSU regions found</td>\n",
       "      <td>139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>586</td>\n",
       "      <td>261</td>\n",
       "      <td>no SSUs were classified</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>587</td>\n",
       "      <td>262</td>\n",
       "      <td>k__Bacteria;p__Bacteroidetes;c__Flavobacteriia...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>588</td>\n",
       "      <td>262</td>\n",
       "      <td>no SSUs were classified</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>589</td>\n",
       "      <td>263</td>\n",
       "      <td>k__Bacteria;p__Bacteroidetes;c__Flavobacteriia...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>590</td>\n",
       "      <td>263</td>\n",
       "      <td>no SSU regions found</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>591 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     group                               SSU_classification_1  sample\n",
       "0        0  k__Eukaryota_(Chloroplast);p__?;c__?;o__?;f__?...       3\n",
       "1        0                               no SSU regions found       1\n",
       "2        0                            no SSUs were classified       1\n",
       "3        1  k__Bacteria;p__Proteobacteria;c__Alphaproteoba...     160\n",
       "4        1                               no SSU regions found     139\n",
       "..     ...                                                ...     ...\n",
       "586    261                            no SSUs were classified       1\n",
       "587    262  k__Bacteria;p__Bacteroidetes;c__Flavobacteriia...       1\n",
       "588    262                            no SSUs were classified       1\n",
       "589    263  k__Bacteria;p__Bacteroidetes;c__Flavobacteriia...       1\n",
       "590    263                               no SSU regions found       1\n",
       "\n",
       "[591 rows x 3 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# just getting at a preview of phylogeny by group membership\n",
    "\n",
    "cdf.groupby(['group','SSU_classification_1'])['sample'].count().reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Output tables..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# I created a possible path for this but feel free to change it.\n",
    "outdir = \"/mnt/scgc/simon/microg2p/Melody/BLM1/Analysis/ANI/tables/\"\n",
    "\n",
    "if not op.exists(outdir):\n",
    "    !mkdir {outdir}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "cdf.to_csv(op.join(outdir, \"ani_{}_cluster_assignments_and_taxonomy.csv\".format(threshold)), index=False)\n",
    "df[df['ani'] >= threshold].to_csv(op.join(outdir, \"ani_{}_edges.csv\".format(threshold)), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Repeating this for several ANI values:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ani_threshold</th>\n",
       "      <th>member_count</th>\n",
       "      <th>cluster_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4086</td>\n",
       "      <td>153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>83</td>\n",
       "      <td>4019</td>\n",
       "      <td>238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>87</td>\n",
       "      <td>3944</td>\n",
       "      <td>264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>93</td>\n",
       "      <td>3796</td>\n",
       "      <td>305</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ani_threshold  member_count  cluster_count\n",
       "0              0          4086            153\n",
       "1             83          4019            238\n",
       "2             87          3944            264\n",
       "3             93          3796            305"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# set up empty dataframe which will collect the cluster information\n",
    "big_cluster_dataframe = pd.DataFrame(columns = ['sample'])\n",
    "\n",
    "# thresholds to use\n",
    "thresholds = [75, 83, 87, 93, 95, 98]\n",
    "\n",
    "# empty lists to collect information about member and cluster counts per threshold\n",
    "members = []\n",
    "clustercount = []\n",
    "\n",
    "\n",
    "for threshold in thresholds:\n",
    "    \n",
    "    # subset fastani table based on threshold\n",
    "    edge_table = df[df['ani'] >= threshold]\n",
    "    \n",
    "    # count how many SAGs are included at given threshold ... there's probably a simpler way to do this\n",
    "    members.append(len(list(set(list(edge_table['genome1'].unique()) + list(edge_table['genome2'].unique())))))\n",
    "    \n",
    "    # cluster the table\n",
    "    clusters = cluster_fastani(df, threshold = threshold)\n",
    "    \n",
    "    # count how many clusters resulted for given threshold\n",
    "    clustercount.append(len(clusters['group'].unique()))\n",
    "    \n",
    "    # change group column name to designate threshold used\n",
    "    clusters = clusters.rename(columns = {'group':'group{}'.format(threshold)})\n",
    "    \n",
    "    # merge cluster table with larger dataframe\n",
    "    big_cluster_dataframe = big_cluster_dataframe.merge(clusters, how='outer')\n",
    "    \n",
    "    # write edge table to csv to be used in cytoscape\n",
    "    edge_table.to_csv(op.join(outdir, \"ani_{}_edges.csv\".format(threshold)), index=False)\n",
    "    \n",
    "    \n",
    "# merge cluster information with taxonomic information    \n",
    "cdf = big_cluster_dataframe.merge(infodf, left_on='sample', right_on='name', how='left')\n",
    "cdf.to_csv(op.join(outdir, \"sag_all_cluster_assignments_and_info.csv\"))\n",
    "\n",
    "# create summary table\n",
    "summarydf = pd.DataFrame(data={'ani_threshold':thresholds,'member_count':members,'cluster_count':clustercount})\n",
    "summarydf.to_csv(op.join(outdir, \"threshold_summary.csv\"))\n",
    "summarydf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "75.6566"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['ani'].min()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0.5, 0, 'ani threshold')"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYsAAAEKCAYAAADjDHn2AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAce0lEQVR4nO3dfZQV9Z3n8fcHcFRc8SG2HkILjSNqjFESblgSHyaDjiHqKEnGCU57xHE2fWQdH88kq8Oso87x5EGzx7AzwemYxKeOxszqaEw0MhpjZsWQbkURRW1UlMgI0TXBdIKK3/2jfj1cmtu3Lth1u6A/r3Pq3Krvrar7vdeSb//qV/UrRQRmZmb1jBruBMzMrPxcLMzMLJeLhZmZ5XKxMDOzXC4WZmaWy8XCzMxyFV4sJI2W9Jiku9PyqZKWS3pXUmXAupdI6pX0jKRPVsWnSVqW3lsgSUXnbWZmmzSjZXE+8HTV8pPAZ4CHqleSdCgwB/ggMAv4hqTR6e2FQAcwJU2zCs7ZzMyqFFosJLUCJwLX9cci4umIeKbG6qcAt0bEhoh4AegFpksaD4yLiMWR3UF4IzC7yLzNzGxzYwre/zXAF4HdG1h3AvBI1fLqFHs7zQ+M17XPPvtEW1tbw4mamRn09PT8KiJaBsYLKxaSTgLWRkSPpE80skmNWNSJ1/rMDrLTVUycOJHu7u4GszUzMwBJq2rFizwNdSRwsqQXgVuBmZJurrP+amD/quVW4JUUb60R30JEdEZEJSIqLS1bFEYzM9tGhRWLiLgkIlojoo2s4/qBiDi9ziZ3AXMk7SxpMllH9pKIWAOslzQjXQV1BnBnUXmbmdmWmn6fhaRPS1oNfAz4oaQfA0TEcuA24CngXuCciNiYNptH1kneC6wE7ml23mZmI5l21CHKK5VKuM/CzGzrSOqJiMrAuO/gNjOzXC4WVbq6oK0NRo3KXru6hjsjM7NyKPo+i+1GVxd0dEBfX7a8alW2DNDePnx5mZmVgVsWyfz5mwpFv76+LG5mNtK5WCQvvbR1cTOzkcTFIpk4ceviZmYjiYtFcuWVMHbs5rGxY7O4mdlI52KRtLdDZydMmgRS9trZ6c5tMzPw1VCbaW93cTAzq8UtCzMzy+ViYWZmuVwszMwsl4uFmZnlcrEwM7NcLhZmZpbLxcLMzHK5WJiZWS4XCzMzy+ViYWZmuVwszMwsl4uFmZnlcrEwM7NcLhZmZpbLxcLMzHK5WJiZWa7Ci4Wk0ZIek3R3Wt5b0iJJz6XXvVK8TdLvJC1N07VV+5gmaZmkXkkLJKnovM3MbJNmtCzOB56uWr4YuD8ipgD3p+V+KyNiaprOroovBDqAKWmaVXDOZmZWpdBiIakVOBG4rip8CnBDmr8BmJ2zj/HAuIhYHBEB3Ji3jZmZDa2iWxbXAF8E3q2K7RcRawDS675V701Op6x+KunoFJsArK5aZ3WKmZlZkxRWLCSdBKyNiJ4GN1kDTIyIDwMXAd+VNA6o1T8Rg3xmh6RuSd3r1q3bprzNzGxLRbYsjgROlvQicCswU9LNwKvp1FL/Kaa1ABGxISJeS/M9wErgILKWRGvVfluBV2p9YER0RkQlIiotLS3FfCszsxGosGIREZdERGtEtAFzgAci4nTgLmBuWm0ucCeApBZJo9P8AWQd2c+nU1XrJc1IV0Gd0b+NmZk1x5hh+MwvA7dJ+ivgJeDUFD8GuELSO8BG4OyIeD29Nw+4HtgVuCdNZmbWJMouMNrxVCqV6O7uHu40zMy2K5J6IqIyMO47uM3MLJeLhZmZ5XKxMDOzXC4WZmaWy8XCzMxyuViYmVkuFwszM8vlYmFmZrlcLMzMLJeLhZmZ5XKxMDOzXC4WZmaWy8XCzMxyuViYNaCrC9raYNSo7LWra7gzMmuu4Xiehdl2pasLOjqgry9bXrUqWwZobx++vMyayS0Lsxzz528qFP36+rK42UjhYmGW46WXti5utiNysTDLMXHi1sXNdkQuFmY5rrwSxo7dPDZ2bBY3GylcLMxytLdDZydMmgRS9trZ6c5tG1l8NZRZA9rbXRxsZHPLwszMcrlYmJlZLhcLMzPL5WJhZma5Ci8WkkZLekzS3Wl5b0mLJD2XXveqWvcSSb2SnpH0yar4NEnL0nsLJKnovM3MbJNmtCzOB56uWr4YuD8ipgD3p2UkHQrMAT4IzAK+IWl02mYh0AFMSdOsJuRtZmZJocVCUitwInBdVfgU4IY0fwMwuyp+a0RsiIgXgF5guqTxwLiIWBwRAdxYtY2ZmTVB0S2La4AvAu9WxfaLiDUA6XXfFJ8AvFy13uoUm5DmB8bNzKxJCisWkk4C1kZET6Ob1IhFnXitz+yQ1C2pe926dQ1+rJmZ5cktFpJ2kzQqzR8k6WRJOzWw7yOBkyW9CNwKzJR0M/BqOrVEel2b1l8N7F+1fSvwSoq31ohvISI6I6ISEZWWlpYGUjQzs0Y00rJ4CNhF0gSyDum/BK7P2ygiLomI1ohoI+u4fiAiTgfuAuam1eYCd6b5u4A5knaWNJmsI3tJOlW1XtKMdBXUGVXbmJlZEzRSLBQRfcBngP8dEZ8GDn0Pn/ll4E8kPQf8SVomIpYDtwFPAfcC50TExrTNPLJO8l5gJXDPe/h8MzPbSo0MJChJHwPagb/aiu3+U0Q8CDyY5l8Djh1kvSuBLQZ+johu4LCt+UwzMxs6jbQszgcuAe6IiOWSDgB+UmxaZmZWJrkthIh4iKzfon/5eeC8IpMyM7NyGbRlIekoSWdULf+LpAfS9MfNSc/MzMqgXsvicuDcquWDgTOB3YC/xaeizMxGjHp9FuMi4qmq5ecioiedltq94LzMzKxE6hWLPasXIuIzVYv7FZOOmZmVUb1isULSiQODaRiPZ4pLyczMyqZen8WFwA8l/RnwaIpNAz4OnFR0YmZmVh6Dtiwiohc4HPgZ0Jamh4DDI+LZZiRnZmblUPc+i4jYAHy7OibpSEl/ERHnFJqZmZmVRkPDdkiaCpwGfA54Abi9yKTMzKxcBi0Wkg4iGy32NOA14Htkgwr6hjwzsxGmXstiBVl/xZ+m/gskXdiUrMzMrFTqXTr7WeA/gJ9I+qakY6n91DozM9vB1bsa6o6I+BxwCNnw4hcC+0laKOn4JuVnZmYlkDtEeUT8NiK6IuIkskeaPgZcXHhmZmZWGvU6uMcCb0fE22n5YOAEYFVEzGxSfmZmVgL1Whb3kt2Ih6QDgcXAAcA5kr5UfGpmZlYW9YrFXhHxXJqfC9wSEecCn8LDfZiZjSj1ikVUzc8EFgFExFvAu0UmZWZm5VLvPosnJF0N/BI4ELgPQNKedbYxM7MdUL2WxeeBX5H1WxwfEX0pfihwdcF5mZlZiQzasoiI3wFfrhF/GHi4yKTMzKxccu+zMDMzc7EwM7NcgxYLSTel1/O3ZceSdpG0RNLjkpZLujzFj5C0WNIyST+QNC7F2yT9TtLSNF1bta9paf1eSQskeYwqM7MmqteymCZpEnCWpL0k7V09NbDvDcDMiDgCmArMkjQDuA64OCI+BNwBfKFqm5URMTVNZ1fFFwIdwJQ0zWr8K5qZ2XtVr1hcS3YX9yFAz4CpO2/HkXkzLe6UpgAOJns8K2T3bny23n4kjQfGRcTiiAjgRmB23uebmdnQqTfq7IKI+ADw7Yg4ICImV00HNLJzSaMlLQXWAosi4ufAk8DJaZVTgf2rNpks6TFJP5V0dIpNAFZXrbM6xczMrEkaGXV2Xupn+Os0Hd7oziNiY0RMJRutdrqkw4CzyMaX6gF2B95Kq68BJkbEh4GLgO+m/oxa/RNRI4akDkndkrrXrVvXaJpmZpYjt1hIOg/oAvZNU5ekc7fmQyLiDbJnYsyKiBURcXxETANuAVamdTZExGtpvifFDyJrSbRW7a4VeGWQz+mMiEpEVFpaWrYmRTPbCl1d0NYGo0Zlr11dw52RFa2RS2f/G/BfI+LSiLgUmEF2d3ddklr6hwaRtCtwHLBC0r4pNgr4O7K+kf71R6f5A8g6sp+PiDXAekkz0lVQZwB3buX3NLMh0tUFHR2wahVEZK8dHS4YO7pGioWAjVXLG2ns8arjyR7J+gTwC7I+i7uB0yQ9S/aM71eA76T1jyEbj+px4F+AsyPi9fTePLKrqHrJWhz3NPD5ZlaA+fOhr2/zWF9fFrcdl7ILjOqsIF1ENkT5HSk0G7g+Iq4pOLf3pFKpRHd37kVbZraVRo3KWhQDSfCux6Pe7knqiYjKwHi9UWcBiIj/JelB4CiyFsVfRsRjQ5+imW0PJk7MTj3VituOK7dYAETEo8CjBediZtuBK6/M+iiqT0WNHZvFbcflsaHMbKu0t0NnJ0yalJ16mjQpW25vH+7MrEgNtSzMzKq1t7s4jDR1WxbpDux/a1YyZmZWTnWLRURsBPok7dGkfMzMrIQaOQ31e2CZpEXAb/uDEXFeYVmZmVmpNFIsfpgmMzMboRq5z+KGNFzHxIh4pgk5mZlZyTQykOCfAkvJnm2BpKmS7io6MTMzK49G7rO4DJgOvAEQEUuByQXmZGZmJdNIsXgnIn49IFZ/QCkzM9uhNNLB/aSkvwBGS5oCnAc8XGxaZmZWJo20LM4FPghsIHtY0W+AC4pMyszMyqWRq6H6gPmSvpItxvri0zIzszJp5Gqoj0paBjxBdnPe45KmFZ+amZmVRSN9Ft8C/ntE/AxA0lFkT7c7vMjEzMysPBrps1jfXygAIuLfAZ+KMjMbQQZtWUj6SJpdIumfyTq3A/gc8GDxqZmZWVnUOw31tQHLf1817/sszMxGkEGLRUT8cTMTMTOz8srt4Ja0J3AG0Fa9vocoNzMbORq5GupHwCPAMuDdYtMxM7MyaqRY7BIRFxWeiZmZlVYjl87eJOnzksZL2rt/KjwzMzMrjUZaFm8BVwHz2XQVVAAHFJWUmZmVSyMti4uAAyOiLSImpym3UEjaRdKSNDzIckmXp/gRkhZLWibpB5LGVW1ziaReSc9I+mRVfFpav1fSAknali9rZmbbppFisRzo24Z9bwBmRsQRwFRglqQZwHXAxRHxIeAO4AsAkg4F5pCNcDsL+Iak0WlfC4EOYEqaZm1DPmZmto0aOQ21EVgq6SdkBQDIv3Q2IgJ4My3ulKYADgYeSvFFwI+B/wmcAtwaERuAFyT1AtMlvQiMi4jFAJJuBGYD9zTyBc3M7L1rpFj8a5q2WmoZ9AAHAv8UET+X9CRwMnAncCqwf1p9Atkluv1Wp9jbaX5gvNbndZC1QJg4ceK2pGxmZjU08jyLG7Z15xGxEZiabuy7Q9JhwFnAAkmXAneRdaAD1OqHiDrxWp/XCXQCVCoVD0liZjZEGrmD+wVq/OPcSCd31bpvSHoQmBURVwPHp30fBJyYVlvNplYGQCvwSoq31oibmVmTNNLBXQE+mqajgQXAzXkbSWpJLQok7QocB6yQtG+KjQL+Drg2bXIXMEfSzpImk3VkL4mINcB6STPSVVBnkJ3CMjOzJsktFhHxWtX0y4i4BpjZwL7HAz+R9ATwC2BRRNwNnCbpWWAFWQvhO+lzlgO3AU8B9wLnpNNYAPPIrqLqBVbizm0zs6ZSdtFSnRU2PdcCsuJSAealS2JLq1KpRHd393CnYWa2XZHUExGVgfFGroaqfq7FO8CLwJ8PUV5mZrYdaORqKD/XwsxshGvkaqidgc+y5fMsriguLTMzK5NGTkPdCfya7Oa6DTnrmpnZDqiRYtEaER6LycxsBGvkPouHJX2o8EzMzKy0GmlZHAWcme7k3kA2/EZExOGFZmZmZqXRSLH4VOFZmJlZqTVy6eyqZiRiZmbl1UifhZmZjXAuFmZmlsvFwszMcrlYmJlZLhcLMzPL5WJhZma5XCzMzCyXi4WZmeVysTAzs1wuFmZmlsvFwszMcrlYmJlZLhcLMzPL5WJhZma5XCzMzCyXi4WZmeUqrFhI2kXSEkmPS1ou6fIUnyrpEUlLJXVLmp7ibZJ+l+JLJV1bta9pkpZJ6pW0QJKKytvMzLbUyGNVt9UGYGZEvClpJ+DfJd0DXAFcHhH3SDoB+CrwibTNyoiYWmNfC4EO4BHgR8As4J4CczczsyqFtSwi82Za3ClNkaZxKb4H8Eq9/UgaD4yLiMUREcCNwOxisjYzs1oK7bOQNFrSUmAtsCgifg5cAFwl6WXgauCSqk0mS3pM0k8lHZ1iE4DVVeusTrFan9eRTm11r1u3bsi/j5nZSFVosYiIjem0UiswXdJhwDzgwojYH7gQ+FZafQ0wMSI+DFwEfFfSOKBW/0QM8nmdEVGJiEpLS8tQfx0zsxGrKVdDRcQbwINkfQ1zgdvTW98Hpqd1NkTEa2m+B1gJHETWkmit2l0rOaeuzMxsaBV5NVSLpD3T/K7AccAKsn/o/yitNhN4rmr90Wn+AGAK8HxErAHWS5qRroI6A7izqLzNzGxLRV4NNR64IRWAUcBtEXG3pDeAr0saA/ye7CongGOAKyS9A2wEzo6I19N784DrgV3JroLylVBmZk2k7AKjHU+lUonu7u7hTsPMbLsiqSciKgPjvoPbzMxyuViYmVkuFwszM8vlYmFmZrlcLMzMLJeLhZmZ5XKxMDOzXC4WZmaWy8XCzMxyuViYmVkuFwszM8vlYmFmZrlcLMzMLJeLhZmZ5XKxMDOzXC4WZmaWy8XCzMxyuViYmVkuFwszM8vlYmFmZrlcLMzMLJeLhZmZ5XKxMDOzXC4WZmaWy8XCzMxyFVYsJO0iaYmkxyUtl3R5ik+V9IikpZK6JU2v2uYSSb2SnpH0yar4NEnL0nsLJKmovM3MbEtFtiw2ADMj4ghgKjBL0gzgq8DlETEVuDQtI+lQYA7wQWAW8A1Jo9O+FgIdwJQ0zSowbzMzG6CwYhGZN9PiTmmKNI1L8T2AV9L8KcCtEbEhIl4AeoHpksYD4yJicUQEcCMwu6i8zcxsS4X2WUgaLWkpsBZYFBE/By4ArpL0MnA1cElafQLwctXmq1NsQpofGK/1eR3p1Fb3unXrhvbLmJmVWFcXtLXBqFHZa1fX0O6/0GIRERvT6aZWslbCYcA84MKI2B+4EPhWWr1WP0TUidf6vM6IqEREpaWl5b1/ATOz7UBXF3R0wKpVEJG9dnQMbcFoytVQEfEG8CBZX8Nc4Pb01veB/g7u1cD+VZu1kp2iWp3mB8bNzAyYPx/6+jaP9fVl8aFS5NVQLZL2TPO7AscBK8j+of+jtNpM4Lk0fxcwR9LOkiaTdWQviYg1wHpJM9JVUGcAdxaVt5nZ9uall7Yuvi3GDN2utjAeuCFd0TQKuC0i7pb0BvB1SWOA35Nd5URELJd0G/AU8A5wTkRsTPuaB1wP7ArckyYzMwMmTsxOPdWKDxVlFxjteCqVSnR3dw93GmZmhevvs6g+FTV2LHR2Qnv71u1LUk9EVAbGfQe3mdl2rr09KwyTJoGUvW5LoainyNNQZmbWJO3tQ1scBnLLwszMcrlYmJlZLhcLMzPL5WJhZma5XCzMzCzXDnufhaR1QI3bVBqyD/CrIUynKM5z6G0vuTrPobe95Fp0npMiYovB9XbYYvFeSOqudVNK2TjPobe95Oo8h972kutw5enTUGZmlsvFwszMcrlY1NY53Ak0yHkOve0lV+c59LaXXIclT/dZmJlZLrcszMws14gqFpIOlrS0avqNpAskXSbpl1XxEwbZfpakZyT1Srp4GPL8XlXsxfR881rbvyhpWVqv0HHaJV0oabmkJyXdImkXSXtLWiTpufS61yDbNuX3rJPnVZJWSHpC0h39D+uqse1w/56lOj5zci3jMXp+ynG5pAtSrIzHaK08y3OMRsSInIDRwH8Ak4DLgL9pYP2VwAHAHwCPA4c2M88B8a8Blw6yzYvAPk3IbQLwArBrWr4NOBP4KnBxil0MfGU4f886eR4PjEmxr9TKsyS/Z+mOz8FyLeExehjwJDCWbJTtfyN7CmfZjtHB8izNMTqiWhYDHAusjIhGb9ybDvRGxPMR8RZwK3BKYdltskWekgT8OXBLEz4/zxhg1/Tkw7Fkj809BbghvX8DMLvGds3+PbfIMyLui4h30vuPsPmz3odLrd+zEcNxfA6aa4mO0Q8Aj0REX/pv/VPg05TvGK2ZZ5mO0ZFcLOaw+YH816mp9+1BmqQTgJerllenWNEG5glwNPBqRDxXY32AAO6T1COpo6jEIuKXwNXAS8Aa4NcRcR+wX2TPTie97ltj86b9nnXyrHYWgz+ud7h/TyjZ8dnAb1qKY5Tsr/VjJL1P0ljgBGB/SnaM1smz2rAeoyOyWEj6A+Bk4PsptBD4Q2Aq2YH/tVqb1YgVeilZjTz7nUb9v9iOjIiPAJ8CzpF0TEH57UX2l9Zk4P3AbpJOb3TzGrFCfs+8PCXNJ3vue9cguxju37N0x2cD/+1LcYxGxNNkp28WAfeSnUp6p+5GmzTtN83LswzH6IgsFmQ/6KMR8SpARLwaERsj4l3gm2TNz4FWs3mlb6XxUwRDkidAavJ/BvjeYBtFxCvpdS1wB7W/z1A4DnghItZFxNvA7cDHgVcljU/5jgfW1ti2mb/nYHkiaS5wEtAe6eTvQMP9e5b0+Kz3m5bpGCUivhURH4mIY4DXgeco3zE6WJ6lOUZHarHY7K+e/oMm+TRZk3CgXwBTJE1Of/HPAe4qNMvaf50dB6yIiNW1NpC0m6Td++fJOshqfZ+h8BIwQ9LYdI76WOBpst9lblpnLnBnjW2b+XvWzFPSLOB/ACdHRF+tDcvwe5b0+Bzsvz2U6xhF0r7pdSJZEbuF8h2jNfMs1TFaZO95GSeyjrjXgD2qYjcBy4AnyA6G8Sn+fuBHVeudADxLdoXE/GbnmeLXA2cPiP1nnmRXbjyepuVNyPNyYEU6OG8CdgbeB9xP9pfR/cDeJfg9a+XZS3ZOemmari3p71m643OwXEt6jP4MeCp93rEpVsZjtFaepTlGfQe3mZnlGqmnoczMbCu4WJiZWS4XCzMzy+ViYWZmuVwszMwsl4uFWQ5JZ0s6o0Z8tqRDq5YflDTkz0ZOI4rusxXrnynpHwd5782hy8xGkjHDnYBZ2UXEtYO8NRu4m+za+IZIGhObBoYz2264ZWEjjqR/TQOuLa8edE3Sm5KulPS4pEck7Zfil0n6mwH7+DjZuF1XpWcI/GF661RJSyQ9K+notO6Zkr4v6QfAfSn2BUm/SIMDXp5iu0n6Yfr8JyV9ruojz5X0qLJnFhyS1t87fZcnUr6H1/iukyUtTp/1D0P2I9qI42JhI9FZETENqADnSXpfiu9GNkz0EcBDwOcH20FEPEx2N/UXImJqRKxMb42JiOnABcDfV23yMWBuRMyUdDzZswqmkw0OOC0N/DaLbOj0IyLiMLIB5fr9KrKB4hYC/YXrcuCxiDgc+Fvgxhqpfh1YGBEfJXsuitk2cbGwkeg8SY+TPR9gf7J/uAHeIjutBNADtG3Dvm8fZPtFEfF6mj8+TY8BjwKHpByWAcdJ+oqkoyPi1zn7PYpsmA0i4gHgfZL2GJDPkWwaX+ymbfg+ZoD7LGyEkfQJsoHuPhYRfZIeBHZJb78dm8a/2ci2/f+xYZDtf1udBvCliPjnGvlNIxuP6EuS7ouIK+rst9EhtD2mj71nblnYSLMH8P9SoTgEmPEe9rUe2H0btvsxcJak/wIgaYKkfSW9H+iLiJvJHiz0kZz9PAS0p318guxU1W8GrPN/yUZLpX9ds23hloWNNPcCZ0t6AniG7FTUtroV+Kak84A/a3SjiLhP0geAxdno3rwJnA4cSNZh/i7wNjAvZ1eXAd9J36WPTUNuVzsf+K6k84H/02iOZgN51FkzM8vl01BmZpbLxcLMzHK5WJiZWS4XCzMzy+ViYWZmuVwszMwsl4uFmZnlcrEwM7Nc/x+waW3LVx9LXgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#substituting a threshold of 75 for 0, because that's the minimum ani reported.\n",
    "\n",
    "plt.plot([75,83,87,93,95,98], summarydf['member_count'], 'ro', color='blue')\n",
    "plt.ylabel('number of SAGs')\n",
    "plt.xlabel('ani threshold')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0.5, 0, 'ani threshold')"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYUAAAEGCAYAAACKB4k+AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAcT0lEQVR4nO3df5QdZZ3n8fcnJILhRwQTWEjS6TgTVoP8brOMiIPEFWQHojOrZoxMWFj6wGSBMOhA6HMGcE6fUdQsuHPUySwgMg0YFtSgjhCyIroCMYmBJgQkTkiMiRBGDWgfQwLf/aOertx07u2u7nTd2z8+r3PuuXWfqqfqeyuV/t6q56mnFBGYmZkBjGl0AGZmNnQ4KZiZWc5JwczMck4KZmaWc1IwM7Pc2EYHsD8mTpwYzc3NjQ7DzGxYWb169csRManavGGdFJqbm1m1alWjwzAzG1Ykbao1z5ePzMws56RgZmY5JwUzM8s5KZiZWc5JwczMck4KZmbDSEdnB803NzPmxjE039xMR2fHoK5/WHdJNTMbTTo6O2h9oJWuXV0AbNqxidYHWgGYd/y8QdmGzxTMzIaJthVteULo1rWri7YVbYO2DScFM7NhYvOOzf0qHwgnBTOzYaJpQlO/ygfCScHMbJhon93O+HHj9yobP2487bPbB20bTgpmZsPEvOPnseS8JUybMA0hpk2YxpLzlgxaIzOAynpGs6SDgEeBA8l6Of2fiLhe0hHA14Fm4AXgoxHxm1RnEXAx8DpwRUQ82Ns2WlpawgPimZn1j6TVEdFSbV6ZZwo7gbMi4kTgJOAcSacB1wIrImIGsCJ9RtJMYC5wHHAO8CVJB5QYn5mZ9VBaUojM79LHcekVwBzgjlR+B/ChND0HuCcidkbERmADMKus+MzMbF+ltilIOkDSWuAlYHlEPAEcFRHbANL7kWnxycAvKqpvSWU919kqaZWkVdu3by8zfDOzUafUpBARr0fEScAUYJakd/ayuKqtoso6l0RES0S0TJpU9cFBZmY2QHXpfRQRvwUeIWsreFHS0QDp/aW02BZgakW1KcDWesRnZmaZ0pKCpEmS3pKm3wy8H3gWWAbMT4vNB76VppcBcyUdKGk6MANYWVZ8Zma2rzIHxDsauCP1IBoDLI2Ib0t6DFgq6WJgM/ARgIhYJ2kp8AywG1gQEa+XGJ+ZmfVQ2n0K9eD7FMzM+q9R9ymYmdkw46RgZmY5JwUzM8s5KZiZWc5JwczMck4KZmaWc1IwM7Ock4KZmeWcFMzMLOekYGZmOScFMzPLOSmYmVnOScHMzHJOCmZmlnNSMDOznJOCmZnlnBTMzCxX5jOap0r6vqT1ktZJujKVnyTpcUlrJa2SNKuiziJJGyQ9J+nssmIzM7PqynxG827g6ohYI+lQYLWk5cBNwI0R8a+Szk2fz5Q0E5gLHAccAzws6Vg/p9nMrH5KO1OIiG0RsSZNvwqsByYDARyWFpsAbE3Tc4B7ImJnRGwENgCzMDOzuinzTCEnqRk4GXgCWAg8KOnzZEnp3WmxycDjFdW2pLKe62oFWgGamppKi9nMbDQqvaFZ0iHAfcDCiHgFuAy4KiKmAlcBt3YvWqV67FMQsSQiWiKiZdKkSWWFbWY2KpWaFCSNI0sIHRFxfyqeD3RP38ueS0RbgKkV1aew59KSmZnVQZm9j0R2FrA+IhZXzNoK/GmaPgt4Pk0vA+ZKOlDSdGAGsLKs+MzMbF9ltimcDlwAdEpam8quAy4BbpE0FvgDqX0gItZJWgo8Q9ZzaYF7HpmZ1VdpSSEifkT1dgKAU2vUaQfay4rJzMx65zuazcws56RgZmY5JwUzM8s5KZiZWc5JwczMck4KZmaWc1IwM7Ock4KZmeWcFMzMLOekYGZmOScFMzPLOSmYmVnOScHMquro7KD55mbG3DiG5pub6ejsaHRIVgd1eRynmQ0vHZ0dtD7QSteuLgA27dhE6wOtAMw7fl4jQ7OS+UzBzPbRtqItTwjdunZ10bairUERWb30mRQknS7p4DT9CUmLJU0rPzQza5TNOzb3q9xGjiJnCl8GuiSdCPwtsAn4WqlRmVlDNU1o6le5jRxFksLuiAhgDnBLRNwCHNpXJUlTJX1f0npJ6yRdWTHvcknPpfKbKsoXSdqQ5p09kC9kZvuvfXY748eN36ts/LjxtM/2gxFHuiINza9KWgR8AnivpAOAcQXq7Qaujog1kg4FVktaDhxFlmBOiIidko4EkDQTmAscBxwDPCzpWD+n2az+uhuT21a0sXnHZpomNNE+u92NzKNAkaTwMeDjwMUR8StJTcDn+qoUEduAbWn6VUnrgcnAJcBnImJnmvdSqjIHuCeVb5S0AZgFPNbP72Rmg2De8fOcBEahXi8fpbOCf4mIxRHxQ4CI2BwR/WpTkNQMnAw8ARwLnCHpCUk/kPSutNhk4BcV1baksp7rapW0StKq7du39ycMMzPrQ69JIV266ZI0YaAbkHQIcB+wMCJeITs7ORw4DfgUsFSSAFULoUpMSyKiJSJaJk2aNNCwzMysiiKXj/4AdKb2gN93F0bEFX1VlDSOLCF0RMT9qXgLcH9qvF4p6Q1gYiqfWlF9CrC10LcwM7NBUSQpfCe9+iX9+r8VWB8RiytmfRM4C3hE0rHAm4CXgWXAXZIWkzU0zwBW9ne7ZmY2cH0mhYi4Q9KbgaaIeK4f6z4duIDsLGNtKrsOuA24TdLTwGvA/HTWsE7SUuAZsp5LC9zzyMysvvpMCpLOAz5P9ot+uqSTgE9HxPm91YuIH1G9nQCy7q3V6rQD7ghtZtYgRW5eu4Gsa+hvASJiLTC9xJjMzKxBit7RvKNH2T69gszMbPgr0tD8tKSPAwdImgFcAfy43LDMzKwRipwpXE429MRO4C5gB3BlrzXMzGxYKnKm8F8iog3IB1KX9BHg3tKiMjOzhihyprCoYJmZmQ1zNc8UJH0QOBeYLOmLFbMOI7uPwMzMRpjeLh9tBVYB5wOrK8pfBa4qMygzM2uMmkkhIp4EnpR0V0TsApB0ODA1In5TrwDNzKx+irQpLJd0mKQjgCeB29P4RGZmNsIUSQoT0pDXfw7cHhGnAu8vNywzM2uEIklhrKSjgY8C3y45HjMza6AiSeHTwIPAhoj4iaS3Ac+XG5aZmTVCkaGz76XiRrWI+DfgL8oMyszMGqPI0Nm3U/2xmBeVEpGZmTVMkWEuKtsRDgI+jB+TaWY2IhW5fHRf5WdJdwMPlxaRmZk1TJGG5p5mAE19LSRpqqTvS1ovaZ2kK3vM/6SkkDSxomyRpA2SnpN09gBiMzOz/VCkTeFVsjYFpfdfAdcUWPdu4OqIWCPpUGC1pOUR8YykqcB/BjZXbGcmMJdsmO5jgIclHevnNJuZ1U+fZwoRcWhEHFbxfmzPS0o16m2LiDVp+lVgPTA5zf6fwN+ydwP2HOCeiNgZERuBDWSPATWrm47ODppvbmbMjWNovrmZjs6ORodkVle9jZJ6Sm8Vu//gFyGpGTgZeELS+cAvI+JJSZWLTQYer/i8hT1JpHJdrUArQFNTn1exzArr6Oyg9YFWunZ1AbBpxyZaH2gFYN7x8xoZmlnd9Hb56Au9zAvgrCIbkHQIcB+wkOySUhvwgWqL1tjO3gURS4AlAC0tLX5WtA2athVteULo1rWri7YVbU4KNmr0Nkrq+/Z35ZLGkSWEjoi4X9LxwHSy0VcBpgBrJM0iOzOYWlF9Cu76anW0ecfmfpWbjUR9tilIWiDpLRWfD5f01wXqCbgVWB8RiwEiojMijoyI5ohoJksEp0TEr4BlwFxJB0qaTtbLaeWAvpXZADRNqH45sla52UhUpEvqJRHx2+4P6VkKlxSodzpwAXCWpLXpdW6thSNiHbAUeAb4HrDAPY+sntpntzN+3Pi9ysaPG0/77PYGRWRWf0XuaB4jSRERAJIOAN7UV6WI+BHV2wkql2nu8bkd8P9Aa4judoO2FW1s3rGZpglNtM9ud3uCjSpFksKDwFJJXyFr+L2U7Je82Ygz7/h5TgI2qhVJCteQdQG9jOyX/0PA/y4zKDMza4wiYx+9AXwlvczMbAQbyNhHZmY2QjkpmJlZrmZSkHRner+y1jJmZjay9HamcKqkacBF6Ya1Iypf9QrQzMzqp7eG5q+QdT19G7Cave85iFRuZmYjSM0zhYj4YkS8A7gtIt4WEdMrXk4IZmYjUJEuqZdJOhE4IxU9GhFPlRuWmZk1QpEB8a4AOoAj06tD0uVlB2ZmZvVX5I7m/w78p4j4PYCkzwKPAf+rzMDMzKz+itynIKBytNLX6WOgOzMzG56KnCncTvYYzW+kzx8ie06CmZmNMEUamhdLegR4D9kZwn+LiJ+WHZiZmdVfkTMFImINsKbkWMzMrME89pGZmeVKSwqSpkr6vqT1ktZ1j6Ek6XOSnpX0lKRv9Hj+8yJJGyQ9J+nssmIzM7Pqek0Kkg6Q9PAA170buDrdFX0asEDSTGA58M6IOAH4GbAobWsmMBc4DjgH+FJ69KeZmdVJr0khIl4HuiRN6O+KI2JbaosgIl4F1gOTI+KhiNidFnscmJKm5wD3RMTOiNgIbABm9Xe7ZmY2cEUamv8AdEpaDvy+uzAirii6EUnNwMnAEz1mXQR8PU1PJksS3baksp7raiV7PChNTU1FQzAzswKKJIXvpNeASDoEuA9YGBGvVJS3kV1i6uguqlI99imIWAIsAWhpadlnvpmZDVyR+xTukPRmoCkinuvPyiWNI0sIHRFxf0X5fODPgNkR0f2HfQswtaL6FGBrf7ZnZmb7p8iAeOcBa8merYCkkyQtK1BPZHc+r4+IxRXl5wDXAOdHRFdFlWXAXEkHSpoOzABW9ufLmJnZ/ily+egGsgbfRwAiYm36o92X04ELyNoj1qay64AvAgcCy7O8weMRcWlErJO0FHiG7LLSgtTQbWZmdVIkKeyOiB3pD3i3Pq/lR8SPqN5O8N1e6rQD7QViMjOzEhRJCk9L+jhwgKQZwBXAj8sNy8zMGqHIHc2Xk91QthO4G3gFWFhmUGZm1hhFeh91AW3p4TqRbkQzM7MRqEjvo3dJ6gSeIms0flLSqeWHZmZm9VakTeFW4K8j4ocAkt5D9uCdE8oMzMzM6q9Im8Kr3QkB8l5FvoRkZjYC1TxTkHRKmlwp6Z/IGpkD+BjpngUzMxtZert89IUen6+vmPaYQ2ZmI1DNpBAR76tnIGZm1nh9NjSnJ6P9FdBcuXx/hs42M7PhoUjvo++SPeegE3ij3HDMzKyRiiSFgyLib0qPxMzMGq5Il9Q7JV0i6WhJR3S/So/MzMzqrsiZwmvA54A29vQ6CuBtZQVlZmaNUSQp/A3wxxHxctnBmJlZYxW5fLQO6OpzKTMzG/aKnCm8DqyV9H2y4bMBd0k1MxuJiiSFb6ZXv0iaCnwN+A9kXVmXRMQtqZH662T3PbwAfDQifpPqLAIuJktEV0TEg/3drpmZDVyR5yncMcB17waujog1kg4FVktaDlwIrIiIz0i6FrgWuEbSTGAu2QN9jgEelnSsn9NsZlY/Re5o3kiVsY4iotfeRxGxDdiWpl+VtB6YDMwBzkyL3UE2uN41qfyeiNgJbJS0AZgFPFbwu5iZ2X4qcvmopWL6IOAjQL/uU5DUDJwMPAEclRIGEbFN0pFpsclkd05325LKeq6rFWgFaGpq6k8YZmbWhz57H0XEv1e8fhkRNwNnFd2ApEOA+4CFEfFKb4tW23yVeJZEREtEtEyaNKloGGZmVkCRy0enVHwcQ3bmcGiRlUsaR5YQOiLi/lT8oqSj01nC0cBLqXwLMLWi+hRga5HtmJnZ4Chy+ajyuQq7ST2G+qokSWSP8lwfEYsrZi0D5gOfSe/fqii/S9JisobmGcDKAvGZmdkgKdL7aKDPVTgduADolLQ2lV1HlgyWSroY2EzWRkFErJO0FHiGLPkscM8jM7P6KnL56EDgL9j3eQqf7q1eepZztXYCgNk16rQD7X3FZGZm5Shy+ehbwA5gNRV3NJuZ2chTJClMiYhzSo/EzMwarsiAeD+WdHzpkZiZWcMVOVN4D3BhurN5J1k7QUTECaVGZmZmdVckKXyw9CjMzGxIKNIldVM9AjEzs8Yr0qZgZmajhJOCmZnlnBTMzCznpGBmZjknBTMzyzkpmJlZzknBzMxyTgpmZpZzUjAzs5yTgpmZ5ZwUzMwsV1pSkHSbpJckPV1RdpKkxyWtlbRK0qyKeYskbZD0nKSzy4rLzMxqK/NM4atAz4fz3ATcGBEnAX+XPiNpJjAXOC7V+ZKkA0qMzczMqigtKUTEo8CvexYDh6XpCcDWND0HuCcidkbERmADMAszM6urIs9TGEwLgQclfZ4sIb07lU8GHq9Ybksq24ekVqAVoKmpqbxIzcxGoXo3NF8GXBURU4GrgFtTuaosG9VWEBFLIqIlIlomTZpUUphmZqNTvZPCfOD+NH0vey4RbQGmViw3hT2XlszMrE7qnRS2An+aps8Cnk/Ty4C5kg6UNB2YAaysc2xmZqNeaW0Kku4GzgQmStoCXA9cAtwiaSzwB1LbQESsk7QUeAbYDSyIiNfLis3MzKorLSlExF/WmHVqjeXbgfay4jEzs775jmYzM8s5KZiZWc5JwczMck4KZmaWc1IwM7Ock4KZmeWcFMzMLOekYGZmOScFMzPLOSmYmVnOScHMzHJOCmZmlnNSMDOznJOCmZnlnBTMzCznpGBmZjknBTMzy5WWFCTdJuklSU/3KL9c0nOS1km6qaJ8kaQNad7ZZcVlZma1lfY4TuCrwD8CX+sukPQ+YA5wQkTslHRkKp8JzAWOA44BHpZ0rJ/TbGZWX6WdKUTEo8CvexRfBnwmInamZV5K5XOAeyJiZ0RsBDYAs8qKzczMqqt3m8KxwBmSnpD0A0nvSuWTgV9ULLclle1DUqukVZJWbd++veRwzcxGl3onhbHA4cBpwKeApZIEqMqyUW0FEbEkIloiomXSpEnlRWpmNgrVOylsAe6PzErgDWBiKp9asdwUYGudYzMzG/XqnRS+CZwFIOlY4E3Ay8AyYK6kAyVNB2YAK+scm5nZqFda7yNJdwNnAhMlbQGuB24DbkvdVF8D5kdEAOskLQWeAXYDC9zzyMys/pT9TR6eWlpaYtWqVY0Ow8xsWJG0OiJaqs0blXc0d3R20HxzM2NuHEPzzc10dHY0OiQzsyGhzJvXhqSOzg5aH2ila1cXAJt2bKL1gVYA5h0/r5GhmZk13Kg7U2hb0ZYnhG5du7poW9HWoIjMzIaOUZcUNu/Y3K9yM7PRZNQlhaYJTf0qNzMbTUZdUmif3c74ceP3Khs/bjzts9sbFJGZ2dAx6pLCvOPnseS8JUybMA0hpk2YxpLzlriR2cwM36dgZjbq+D4FMzMrxEnBzMxyTgpmZpZzUjAzs5yTgpmZ5YZ17yNJ24FN+7GKiWTPcxjqHOfgGi5xwvCJ1XEOrrLjnBYRVR9dOayTwv6StKpWt6yhxHEOruESJwyfWB3n4GpknL58ZGZmOScFMzPLjfaksKTRARTkOAfXcIkThk+sjnNwNSzOUd2mYGZmexvtZwpmZlbBScHMzHIjLilI+o+S1la8XpG0UNINkn5ZUX5ujfrnSHpO0gZJ1zYgzq9XlL0gaW2N+i9I6kzLlT5UrKSrJK2T9LSkuyUdJOkIScslPZ/eD69Rty77tJc4PyfpWUlPSfqGpLfUqFu3fVojziF1jPYS51A9Rq9Mca6TtDCVDcVjtFqcQ+cYjYgR+wIOAH4FTANuAD5ZYPmfA28D3gQ8CcysZ5w9yr8A/F2NOi8AE+u0HycDG4E3p89LgQuBm4BrU9m1wGcbuU97ifMDwNhU9tlqcdZzn/YS55A6RmvFOUSP0XcCTwPjgbHAw8CMIXiM1opzyByjI+5MoYfZwM8jouhdz7OADRHxbxHxGnAPMKe06PbYJ05JAj4K3F2H7RcxFnizpLFkB/RWsn1zR5p/B/ChKvXqvU/3iTMiHoqI3Wn+48CUErdfVLX9WUTD92f3jCF2jL4DeDwiutK/9Q+ADzP0jtGqcQ6lY3SkJ4W57H3A/o90enZbjdPIycAvKj5vSWVl6xknwBnAixHxfI06ATwkabWk1jKDi4hfAp8HNgPbgB0R8RBwVERsS8tsA46sUr1u+7SXOCtdBPxrrVVQh33aR5xD5hgtsD+HzDFK9uv7vZLeKmk8cC4wlSF2jPYSZ6WGHqMjNilIehNwPnBvKvoy8EfASWQH+BeqVatSVmqf3SpxdvtLev8FdnpEnAJ8EFgg6b0lhUj64zQHmA4cAxws6RNFq1cpK2Wf9hWnpDZgN9BRYxV12ae9xDmkjtEC/+5D5hiNiPVkl12WA98juwS0u9dKe9Rtn/YV51A4RkdsUiDbaWsi4kWAiHgxIl6PiDeAfyY7ZexpC3tn7SkUP60flDgB0qn6nwNfr1UpIram95eAb1D9+wyW9wMbI2J7ROwC7gfeDbwo6egU89HAS1Xq1nOf1ooTSfOBPwPmRbo421Md92nVOIfgMdrb/hxqxygRcWtEnBIR7wV+DTzP0DtGa8U5ZI7RkZwU9voV031gJB8mO43r6SfADEnT0y/4ucCyUqOs/mvr/cCzEbGlWgVJB0s6tHuarJGq2vcZLJuB0ySNT9eRZwPryfbN/LTMfOBbVerWc59WjVPSOcA1wPkR0VWtYp33aa04h9oxWuvfHYbeMYqkI9N7E1nCupuhd4xWjXNIHaNltmI36kXWIPbvwISKsjuBTuApsn/wo1P5McB3K5Y7F/gZWW+EtnrHmcq/ClzaoyyPk6yXxJPpta7sONM2bwSeTQfhncCBwFuBFWS/dFYARwyBfVotzg1k14zXptdXGr1Pa8Q5FI/RfeIcwsfoD4Fn0jZnp7KheIxWi3PIHKMe5sLMzHIj+fKRmZn1k5OCmZnlnBTMzCznpGBmZjknBTMzyzkpmCWSLpX0V1XKPyRpZsXnRyQN+kPV0wiYE/ux/IWS/rHGvN8NXmQ2moxtdABmQ0VEfKXGrA8B3ybrW16IpLGxZ4Azs2HDZwo2Ykn6Zho4bF3l4GGSfiepXdKTkh6XdFQqv0HSJ3us491kY1N9Lo1h/0dp1kckrZT0M0lnpGUvlHSvpAeAh1LZpyT9JA1yd2MqO1jSd9L2n5b0sYpNXi5pjbIx89+elj8ifZenUrwnVPmu0yU9lrb194O2E23UcVKwkeyiiDgVaAGukPTWVH4w2fDFJwKPApfUWkFE/Jjs7uJPRcRJEfHzNGtsRMwCFgLXV1T5E2B+RJwl6QNkY+XPIhvk7tQ0gNk5ZEN6nxgR7yQbGK3by5ENePZloDtB3Qj8NCJOAK4DvlYl1FuAL0fEu8iezWE2IE4KNpJdIelJsvHpp5L9gQZ4jexyEMBqoHkA676/Rv3lEfHrNP2B9PopsAZ4e4qhE3i/pM9KOiMidvSx3veQDTFBRPxf4K2SJvSI53T2jKF15wC+jxngNgUboSSdSTZo259ERJekR4CD0uxdsWd8l9cZ2P+DnTXq/74yDOAfIuKfqsR3Ktl4O/8g6aGI+HQv6y06tLPHrLH95jMFG6kmAL9JCeHtwGn7sa5XgUMHUO9B4CJJhwBImizpSEnHAF0R8S9kD7E5pY/1PArMS+s4k+wS0ys9lvl/ZKN70r2s2UD4TMFGqu8Bl0p6CniO7BLSQN0D/LOkK4D/WrRSRDwk6R3AY9nI0/wO+ATwx2QN128Au4DL+ljVDcDt6bt0sWco6EpXAndJuhK4r2iMZj15lFQzM8v58pGZmeWcFMzMLOekYGZmOScFMzPLOSmYmVnOScHMzHJOCmZmlvv/uhlmlO7XEnoAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.plot([75,83,87,93,95,98], summarydf['cluster_count'], 'ro', color='green')\n",
    "plt.ylabel('number of clusters')\n",
    "plt.xlabel('ani threshold')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Before we get into categorizing SAGs, we need to address contamination, and determine what ani threshold we'll want to use.\n",
    "\n",
    "Regarding contamination, a quick look at the networks with cytoscape suggests we still have to clean up some SAGs that might be co-sorts or otherwise contaminated."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First I want to find out how many network clusters have more than one phylogenetic group in them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# change the 'no SSU...' classifications to 'None' for easier interpretation\n",
    "\n",
    "### SAGs that are unclassified via GTDB might need to be changed as well. CHANGE the .str.contains ('**') to something in all unclassified GTDB assignments\n",
    "\n",
    "\n",
    "cdf.loc[cdf['SSU_classification_1'].str.contains('no SSU'),'SSU_classification_1'] = 'None'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# break up classifications\n",
    "\n",
    "### I think that this will work but This might need to be modified if the GTDB_classificaiton is using a different \n",
    "### seperator than the original SSU column was to deliminate taxonomic levels. \n",
    "### replace the \"__\" with whatever the speerator for GTDB is \"**\"\n",
    "\n",
    "cdf['classlist'] = [i.split(\";\") for i in cdf['SSU_classification_1']]\n",
    "cdf['kingdom'] = [i[0].split(\"__\")[-1] for i in cdf['classlist']]\n",
    "cdf['phylum'] = [i[1].split(\"__\")[-1] if len(i) > 1 else 'None' for i in cdf['classlist']]\n",
    "cdf['class'] = [i[2].split(\"__\")[-1] if len(i) > 2 else 'None' for i in cdf['classlist']]\n",
    "cdf['order'] = [i[3].split(\"__\")[-1] if len(i) > 3 else 'None' for i in cdf['classlist']]\n",
    "cdf['family'] = [i[4].split(\"__\")[-1] if len(i) > 4 else 'None' for i in cdf['classlist']]\n",
    "cdf['genus'] = [i[5].split(\"__\")[-1] if len(i) > 5 else 'None' for i in cdf['classlist']]\n",
    "cdf['species'] = [i[6].split(\"__\")[-1] if len(i) > 6 else 'None' for i in cdf['classlist']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>group0</th>\n",
       "      <th>group83</th>\n",
       "      <th>group87</th>\n",
       "      <th>group93</th>\n",
       "      <th>name</th>\n",
       "      <th>Plate</th>\n",
       "      <th>SSU_classification_1</th>\n",
       "      <th>well_type</th>\n",
       "      <th>classlist</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>AH-700-O11</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AH-700-O11</td>\n",
       "      <td>AH-700</td>\n",
       "      <td>k__Eukaryota_(Chloroplast);p__?;c__?;o__?;f__?...</td>\n",
       "      <td>1 cell</td>\n",
       "      <td>[k__Eukaryota_(Chloroplast), p__?, c__?, o__?,...</td>\n",
       "      <td>Eukaryota_(Chloroplast)</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>AH-709-G13</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>AH-709-G13</td>\n",
       "      <td>AH-709</td>\n",
       "      <td>None</td>\n",
       "      <td>1 cell</td>\n",
       "      <td>[None]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>AH-578-K18</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>AH-578-K18</td>\n",
       "      <td>AH-578</td>\n",
       "      <td>k__Bacteria;p__Bacteroidetes;c__Sphingobacteri...</td>\n",
       "      <td>1 cell</td>\n",
       "      <td>[k__Bacteria, p__Bacteroidetes, c__Sphingobact...</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>Sphingobacteriia</td>\n",
       "      <td>Sphingobacteriales</td>\n",
       "      <td>NS11-12_marine_group</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>AH-578-M09</td>\n",
       "      <td>3</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>AH-578-M09</td>\n",
       "      <td>AH-578</td>\n",
       "      <td>k__Bacteria;p__Bacteroidetes;c__Flavobacteriia...</td>\n",
       "      <td>1 cell</td>\n",
       "      <td>[k__Bacteria, p__Bacteroidetes, c__Flavobacter...</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>Flavobacteriia</td>\n",
       "      <td>Flavobacteriales</td>\n",
       "      <td>Crocinitomicaceae</td>\n",
       "      <td>Fluviicola</td>\n",
       "      <td>?</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>AH-700-A04</td>\n",
       "      <td>4</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>AH-700-A04</td>\n",
       "      <td>AH-700</td>\n",
       "      <td>None</td>\n",
       "      <td>1 cell</td>\n",
       "      <td>[None]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       sample  group0  group83  group87  group93        name   Plate  \\\n",
       "0  AH-700-O11       0      0.0      0.0      NaN  AH-700-O11  AH-700   \n",
       "1  AH-709-G13       1      1.0      1.0      0.0  AH-709-G13  AH-709   \n",
       "2  AH-578-K18       2      2.0      2.0      1.0  AH-578-K18  AH-578   \n",
       "3  AH-578-M09       3      3.0      3.0      2.0  AH-578-M09  AH-578   \n",
       "4  AH-700-A04       4      4.0      4.0      3.0  AH-700-A04  AH-700   \n",
       "\n",
       "                                SSU_classification_1 well_type  \\\n",
       "0  k__Eukaryota_(Chloroplast);p__?;c__?;o__?;f__?...    1 cell   \n",
       "1                                               None    1 cell   \n",
       "2  k__Bacteria;p__Bacteroidetes;c__Sphingobacteri...    1 cell   \n",
       "3  k__Bacteria;p__Bacteroidetes;c__Flavobacteriia...    1 cell   \n",
       "4                                               None    1 cell   \n",
       "\n",
       "                                           classlist                  kingdom  \\\n",
       "0  [k__Eukaryota_(Chloroplast), p__?, c__?, o__?,...  Eukaryota_(Chloroplast)   \n",
       "1                                             [None]                     None   \n",
       "2  [k__Bacteria, p__Bacteroidetes, c__Sphingobact...                 Bacteria   \n",
       "3  [k__Bacteria, p__Bacteroidetes, c__Flavobacter...                 Bacteria   \n",
       "4                                             [None]                     None   \n",
       "\n",
       "          phylum             class               order                family  \\\n",
       "0              ?                 ?                   ?                     ?   \n",
       "1           None              None                None                  None   \n",
       "2  Bacteroidetes  Sphingobacteriia  Sphingobacteriales  NS11-12_marine_group   \n",
       "3  Bacteroidetes    Flavobacteriia    Flavobacteriales     Crocinitomicaceae   \n",
       "4           None              None                None                  None   \n",
       "\n",
       "        genus species  \n",
       "0           ?       ?  \n",
       "1        None    None  \n",
       "2           ?       ?  \n",
       "3  Fluviicola       ?  \n",
       "4        None    None  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cdf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set group to examine. The # is equal to the threshold eg group95 is 95%ANI\n",
    "group = 'group0'\n",
    "\n",
    "# count group membership by taxonomy (down to family level here)\n",
    "group_tax_counts = cdf.groupby([group, 'phylum','class','order','family'])['sample'].count().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group0</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>?</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>SAR11_clade</td>\n",
       "      <td>?</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>SAR11_clade</td>\n",
       "      <td>Surface_1</td>\n",
       "      <td>309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>SAR11_clade</td>\n",
       "      <td>Surface_2</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>Sphingobacteriia</td>\n",
       "      <td>Sphingobacteriales</td>\n",
       "      <td>NS11-12_marine_group</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>Flavobacteriia</td>\n",
       "      <td>Flavobacteriales</td>\n",
       "      <td>Crocinitomicaceae</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10</td>\n",
       "      <td>4</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>11</td>\n",
       "      <td>5</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>12</td>\n",
       "      <td>5</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rickettsiales</td>\n",
       "      <td>SAR116_clade</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>13</td>\n",
       "      <td>6</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>14</td>\n",
       "      <td>6</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Oceanospirillales</td>\n",
       "      <td>SAR86_clade</td>\n",
       "      <td>101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15</td>\n",
       "      <td>7</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>16</td>\n",
       "      <td>7</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>SAR11_clade</td>\n",
       "      <td>?</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>17</td>\n",
       "      <td>7</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>SAR11_clade</td>\n",
       "      <td>Chesapeake-Delaware_Bay</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18</td>\n",
       "      <td>8</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>19</td>\n",
       "      <td>8</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rhodobacterales</td>\n",
       "      <td>Rhodobacteraceae</td>\n",
       "      <td>114</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    group0          phylum                class               order  \\\n",
       "0        0               ?                    ?                   ?   \n",
       "1        0            None                 None                None   \n",
       "2        1            None                 None                None   \n",
       "3        1  Proteobacteria  Alphaproteobacteria         SAR11_clade   \n",
       "4        1  Proteobacteria  Alphaproteobacteria         SAR11_clade   \n",
       "5        1  Proteobacteria  Alphaproteobacteria         SAR11_clade   \n",
       "6        2   Bacteroidetes     Sphingobacteriia  Sphingobacteriales   \n",
       "7        2            None                 None                None   \n",
       "8        3   Bacteroidetes       Flavobacteriia    Flavobacteriales   \n",
       "9        3            None                 None                None   \n",
       "10       4            None                 None                None   \n",
       "11       5            None                 None                None   \n",
       "12       5  Proteobacteria  Alphaproteobacteria       Rickettsiales   \n",
       "13       6            None                 None                None   \n",
       "14       6  Proteobacteria  Gammaproteobacteria   Oceanospirillales   \n",
       "15       7            None                 None                None   \n",
       "16       7  Proteobacteria  Alphaproteobacteria         SAR11_clade   \n",
       "17       7  Proteobacteria  Alphaproteobacteria         SAR11_clade   \n",
       "18       8            None                 None                None   \n",
       "19       8  Proteobacteria  Alphaproteobacteria     Rhodobacterales   \n",
       "\n",
       "                     family  sample  \n",
       "0                         ?       3  \n",
       "1                      None       2  \n",
       "2                      None     274  \n",
       "3                         ?       3  \n",
       "4                 Surface_1     309  \n",
       "5                 Surface_2       5  \n",
       "6      NS11-12_marine_group       8  \n",
       "7                      None       5  \n",
       "8         Crocinitomicaceae       8  \n",
       "9                      None       5  \n",
       "10                     None       3  \n",
       "11                     None      27  \n",
       "12             SAR116_clade      32  \n",
       "13                     None      90  \n",
       "14              SAR86_clade     101  \n",
       "15                     None       7  \n",
       "16                        ?       3  \n",
       "17  Chesapeake-Delaware_Bay       6  \n",
       "18                     None      71  \n",
       "19         Rhodobacteraceae     114  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check result\n",
    "\n",
    "group_tax_counts[:20]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group0</th>\n",
       "      <th>family</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>68</td>\n",
       "      <td>74</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>34</td>\n",
       "      <td>36</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>89</td>\n",
       "      <td>99</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>85</td>\n",
       "      <td>94</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>86</td>\n",
       "      <td>95</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>87</td>\n",
       "      <td>96</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>88</td>\n",
       "      <td>98</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    group0  family\n",
       "1        1       3\n",
       "68      74       2\n",
       "34      36       2\n",
       "6        7       2\n",
       "0        0       1\n",
       "89      99       1\n",
       "85      94       1\n",
       "86      95       1\n",
       "87      96       1\n",
       "88      98       1"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# remove 'none' category and count families per group\n",
    "# This is a contamination check.\n",
    "family_by_group = group_tax_counts[group_tax_counts['family'] != 'None'].groupby(group)['family'].count().reset_index()\n",
    "\n",
    "family_by_group.sort_values(by='family', ascending=False)[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Only a couple of the groups have differing family memberships... something to look into.  \n",
    "\n",
    "Let's look at a higher tax levels to see how the result changes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group0</th>\n",
       "      <th>order</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>68</td>\n",
       "      <td>74</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>34</td>\n",
       "      <td>36</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>89</td>\n",
       "      <td>99</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>85</td>\n",
       "      <td>94</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>86</td>\n",
       "      <td>95</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>87</td>\n",
       "      <td>96</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>88</td>\n",
       "      <td>98</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    group0  order\n",
       "1        1      3\n",
       "68      74      2\n",
       "34      36      2\n",
       "6        7      2\n",
       "0        0      1\n",
       "89      99      1\n",
       "85      94      1\n",
       "86      95      1\n",
       "87      96      1\n",
       "88      98      1"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "level = 'order'\n",
    "\n",
    "count_by_group = group_tax_counts[group_tax_counts[level] != 'None'].groupby(group)[level].count().reset_index()\n",
    "\n",
    "count_by_group.sort_values(by=level, ascending=False)[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Same result."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group0</th>\n",
       "      <th>class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>68</td>\n",
       "      <td>74</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>34</td>\n",
       "      <td>36</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>89</td>\n",
       "      <td>99</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>85</td>\n",
       "      <td>94</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>86</td>\n",
       "      <td>95</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>87</td>\n",
       "      <td>96</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>88</td>\n",
       "      <td>98</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    group0  class\n",
       "1        1      3\n",
       "68      74      2\n",
       "34      36      2\n",
       "6        7      2\n",
       "0        0      1\n",
       "89      99      1\n",
       "85      94      1\n",
       "86      95      1\n",
       "87      96      1\n",
       "88      98      1"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "level = 'class'\n",
    "\n",
    "count_by_group = group_tax_counts[group_tax_counts[level] != 'None'].groupby(group)[level].count().reset_index()\n",
    "\n",
    "count_by_group.sort_values(by=level, ascending=False)[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Same result."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "problematic_clusters = list(count_by_group[count_by_group[level] > 1 ][group])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "clusters to check for contamination: [1, 7, 36, 74]\n"
     ]
    }
   ],
   "source": [
    "print(\"clusters to check for contamination:\", problematic_clusters)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Also problematic will be nodes that connect two clusters together... these could potentially be co-sorts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "invalid literal for int() with base 10: 'p0'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-31-7d01edb04a93>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# add group membership to edge table\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mthreshold\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgroup\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Looking into contaminination using this threshold\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mthreshold\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: invalid literal for int() with base 10: 'p0'"
     ]
    }
   ],
   "source": [
    "# add group membership to edge table\n",
    "\n",
    "threshold = int(group[-2:])\n",
    "print(\"Looking into contaminination using this threshold\", threshold)\n",
    "\n",
    "edgedf = df[df['ani'] >= threshold]\n",
    "\n",
    "edgedf = edgedf.merge(cdf[[\"sample\",group]].rename(columns={'sample':'genome1', group:'genome1_group'}), on='genome1', how='left').\\\n",
    "merge(cdf[[\"sample\",group]].rename(columns={'sample':'genome2', group:'genome2_group'}), on='genome2', how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "conflict_edges = edgedf[edgedf['genome1_group'] != edgedf['genome2_group']].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "group_pairs = []\n",
    "\n",
    "for i, l in conflict_edges.iterrows():\n",
    "    pair = \"_\".join(sorted([str(l['genome1_group']), str(l['genome2_group'])]))\n",
    "    group_pairs.append(pair)\n",
    "    \n",
    "conflict_edges['group_pairs'] = group_pairs\n",
    "\n",
    "contam_scan = pd.concat([conflict_edges[['genome1','group_pairs']].rename(columns={'genome1':'genome'}),\n",
    "         conflict_edges[['genome2','group_pairs']].rename(columns={'genome1':'genome'})]).groupby('genome')['group_pairs'].count().reset_index()\n",
    "\n",
    "contam_scan[contam_scan['group_pairs'] > 1].sort_values(by='group_pairs', ascending=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "AH-299-B13 has lots of connections to nodes from a different group, making it alikely culprit to be a contaminant.  The other two nodes are more ambiguous."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next Steps would be to investigate the network using a visual tool like cytoscape and examine the problematic clusters and SAGs to determine if they are contaminated and need to be removed.  Then we can re-run this ANI analysis with a cleaner set of SAGs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
