{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8ededb86-1cd6-4895-9c80-a4abffaa0f82",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Folder already exists\n"
     ]
    }
   ],
   "source": [
    "import glob\n",
    "from collections import defaultdict\n",
    "import pandas as pd\n",
    "import os.path as op\n",
    "import math\n",
    "import os \n",
    "import sys\n",
    "import numpy as np\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from collections import Counter\n",
    "\n",
    "import sqlite3\n",
    "import sqlalchemy\n",
    "\n",
    "def safe_makedir(dname):\n",
    "    \"\"\"\n",
    "    Make a directory if it doesn't exist, handling concurrent race conditions.\n",
    "    \"\"\"\n",
    "    if not dname:\n",
    "        return dname\n",
    "    num_tries = 0\n",
    "    max_tries = 5\n",
    "    while not os.path.exists(dname):\n",
    "        try:\n",
    "            os.makedirs(dname)\n",
    "        except OSError:\n",
    "            if num_tries > max_tries:\n",
    "                raise\n",
    "            num_tries += 1\n",
    "            time.sleep(2)\n",
    "    return dname\n",
    "\n",
    "# import my custom functions\n",
    "sys.path.insert(0, '/home/jmunson-mcgee/')\n",
    "from JMM_functions import *\n",
    "\n",
    "pd.set_option(\"display.max_rows\", 6)\n",
    "\n",
    "outdir='/mnt/scgc/simon/microg2p/Melody/BLM1/Analyses/RNA_GORG_recruit/Summary_files/'\n",
    "safe_make_dir(outdir)\n",
    "os.chdir(outdir)\n",
    "\n",
    "    \n",
    "matplotlib.__version__\n",
    "\n",
    "read_files = glob.glob('/mnt/scgc/simon/microg2p/Melody/BLM1/Analyses/RNA_GORG_recruit/Summary_files/*annotated_RNA_GORG_results.csv')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "374f881a-a21d-4a7e-98aa-74be8bf07ff8",
   "metadata": {},
   "source": [
    "# First steps\n",
    "1. Use the 10none and 500none recruitments to calculate how many cells/ml there are for genera/families with at least 5 cells that are 20% complete or more\n",
    "1. Calculate the number of reads per gene for each sample (this is probably more of an RNA thing but oh well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "748be2eb-d3ce-4578-8af0-08011d448402",
   "metadata": {},
   "outputs": [],
   "source": [
    "SAGdata='/mnt/scgc/simon/microg2p/analyses/JdF_analysis/summary_files/JdF_SAGs_GTDB_2.0_and_95ANI_clustering_master_predresp_rate.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "13cfc72f-ed0f-4401-8b4c-54da1ae6f61b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cols=['name', 'GTDB_classification', 'checkM_estimated_completeness', 'hypothetical_cds_fraction', 'genus', 'family', 'depth']\n",
    "\n",
    "# JdF_SAGs=pd.read_csv(SAGdata)#, usecols=cols)\n",
    "# JdF_SAGs.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e5a40489-137c-46f9-9dcb-d03849d6a6f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "168\n",
      "99\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>family</th>\n",
       "      <th>hypothetical_cds_fraction</th>\n",
       "      <th>checkM_estimated_completeness</th>\n",
       "      <th>final_assembly_length</th>\n",
       "      <th>number_of_cells</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2-12-FULL-41-12</td>\n",
       "      <td>0.397177</td>\n",
       "      <td>43.180000</td>\n",
       "      <td>5.016050e+05</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAA536-G10</td>\n",
       "      <td>0.244515</td>\n",
       "      <td>27.107222</td>\n",
       "      <td>6.986373e+05</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AG-339-G14</td>\n",
       "      <td>0.233311</td>\n",
       "      <td>36.849000</td>\n",
       "      <td>6.941456e+05</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>UBA890</td>\n",
       "      <td>0.364000</td>\n",
       "      <td>10.340000</td>\n",
       "      <td>5.349430e+05</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>UBA9320</td>\n",
       "      <td>0.342016</td>\n",
       "      <td>44.830000</td>\n",
       "      <td>1.026203e+06</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>Woeseiaceae</td>\n",
       "      <td>0.251808</td>\n",
       "      <td>16.718000</td>\n",
       "      <td>5.169096e+05</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>99 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             family  hypothetical_cds_fraction  checkM_estimated_completeness  \\\n",
       "0   2-12-FULL-41-12                   0.397177                      43.180000   \n",
       "1        AAA536-G10                   0.244515                      27.107222   \n",
       "2        AG-339-G14                   0.233311                      36.849000   \n",
       "..              ...                        ...                            ...   \n",
       "96           UBA890                   0.364000                      10.340000   \n",
       "97          UBA9320                   0.342016                      44.830000   \n",
       "98      Woeseiaceae                   0.251808                      16.718000   \n",
       "\n",
       "    final_assembly_length  number_of_cells  \n",
       "0            5.016050e+05                1  \n",
       "1            6.986373e+05               36  \n",
       "2            6.941456e+05               10  \n",
       "..                    ...              ...  \n",
       "96           5.349430e+05                1  \n",
       "97           1.026203e+06                1  \n",
       "98           5.169096e+05                5  \n",
       "\n",
       "[99 rows x 5 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cols=['name', 'GTDB_classification', 'final_assembly_length', 'checkM_estimated_completeness', 'hypothetical_cds_fraction', 'genus', 'family', 'depth']\n",
    "\n",
    "JdF_SAGs=pd.read_csv(SAGdata)#, usecols=cols)\n",
    "# JdF_SAGs=JdF_SAGs[JdF_SAGs['checkM_estimated_completeness']>=20]\n",
    "\n",
    "JdF_genus_counts=JdF_SAGs.groupby('genus', as_index=False)['hypothetical_cds_fraction'].count()\n",
    "JdF_genus_counts.rename(columns = {'hypothetical_cds_fraction':'number_of_cells'}, inplace=True)\n",
    "\n",
    "JdF_genus_stats=JdF_SAGs.groupby('genus', as_index=False)[['hypothetical_cds_fraction', 'checkM_estimated_completeness', 'final_assembly_length']].mean()\n",
    "JdF_genus_stats=JdF_genus_stats.merge(JdF_genus_counts, on='genus', how='left')\n",
    "\n",
    "JdF_genus_stats['genus']=JdF_genus_stats['genus'].str.strip()\n",
    "print(len(JdF_genus_counts))\n",
    "\n",
    "\n",
    "\n",
    "JdF_family_counts=JdF_SAGs.groupby('family', as_index=False)['hypothetical_cds_fraction'].count()\n",
    "JdF_family_counts.rename(columns = {'hypothetical_cds_fraction':'number_of_cells'}, inplace=True)\n",
    "\n",
    "JdF_family_stats=JdF_SAGs.groupby('family', as_index=False)[['hypothetical_cds_fraction', 'checkM_estimated_completeness', 'final_assembly_length']].mean()\n",
    "JdF_family_stats=JdF_family_stats.merge(JdF_family_counts, on='family', how='left')\n",
    "\n",
    "JdF_family_stats['family']=JdF_family_counts['family'].str.strip()\n",
    "# JdF_family_counts=JdF_family_counts[JdF_family_counts['number_of_cells']>=5]\n",
    "\n",
    "\n",
    "print(len(JdF_family_stats))\n",
    "JdF_family_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5caee6a-8933-42c9-b096-b736e653554e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "69e80e32-1d2d-4e72-b126-52fd8e5a3f35",
   "metadata": {},
   "source": [
    "This leaves me with 50 genera and 36 families (including Unclassified) that have at least 5 cells with a checkm completeness of at least 20%. \n",
    "\n",
    "\n",
    "I still need to find my file that has the total number of reads per sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "22e88f96-d3aa-4c54-9cb8-429d4df328c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This step takes a while so I ran it once and then hardocded the output into the variables for the number of reads and comment out the count commands\n",
    "# ! zcat /mnt/scgc/simon/microg2p/JdF_analysis/analysis/JdF_GORG_RNA_recruit/annotations/All_10none_reads_noMTST_annotated.txt.gz | wc -l\n",
    "# ! zcat /mnt/scgc/simon/microg2p/JdF_analysis/analysis/JdF_GORG_RNA_recruit/annotations/All_500none_reads_noMTST_annotated.txt.gz | wc -l\n",
    "\n",
    "\n",
    "All10_reads=106908161\n",
    "All500_reads=84370726\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "1a81ff33-8be9-4ae7-809e-7665f3afdd09",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cDNA4-A01_joined_annotated_RNA_GORG_\n",
      "cDNA2-A01_joined_annotated_RNA_GORG_\n",
      "cDNA3-A01_joined_annotated_RNA_GORG_\n",
      "cDNA7-A01_joined_annotated_RNA_GORG_\n",
      "cDNA1-A01_joined_annotated_RNA_GORG_\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>order</th>\n",
       "      <th>Aminicenantales</th>\n",
       "      <th>B26-1</th>\n",
       "      <th>Bryobacterales</th>\n",
       "      <th>Carboxydocellales</th>\n",
       "      <th>Hadarchaeales</th>\n",
       "      <th>Methanomassiliicoccales</th>\n",
       "      <th>NA</th>\n",
       "      <th>RBG-16-55-9</th>\n",
       "      <th>Thermodesulfovibrionales</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_cluster_rep</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AM-294-J06_N23;5746;6825</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AM-294-K13_N4;7685;8899</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>46</td>\n",
       "      <td>0</td>\n",
       "      <td>18569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AM-294-L07_N4;47996;48265</th>\n",
       "      <td>0</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AM-302-A23_N6;3581;4708</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>46</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AM-304-B23_N71;959;1417</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AM-307-C17_N9;25035;25322</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>24331</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "order                       Aminicenantales   B26-1   Bryobacterales  \\\n",
       "gene_cluster_rep                                                       \n",
       "AM-294-J06_N23;5746;6825                  0       0                0   \n",
       "AM-294-K13_N4;7685;8899                   0       0                0   \n",
       "AM-294-L07_N4;47996;48265                 0      15                0   \n",
       "...                                     ...     ...              ...   \n",
       "AM-302-A23_N6;3581;4708                   0       0                0   \n",
       "AM-304-B23_N71;959;1417                   0       0                0   \n",
       "AM-307-C17_N9;25035;25322                 0       0                0   \n",
       "\n",
       "order                       Carboxydocellales   Hadarchaeales  \\\n",
       "gene_cluster_rep                                                \n",
       "AM-294-J06_N23;5746;6825                    1               0   \n",
       "AM-294-K13_N4;7685;8899                     0               0   \n",
       "AM-294-L07_N4;47996;48265                   0               0   \n",
       "...                                       ...             ...   \n",
       "AM-302-A23_N6;3581;4708                     0               0   \n",
       "AM-304-B23_N71;959;1417                     0               0   \n",
       "AM-307-C17_N9;25035;25322                   0               0   \n",
       "\n",
       "order                       Methanomassiliicoccales   NA   RBG-16-55-9  \\\n",
       "gene_cluster_rep                                                         \n",
       "AM-294-J06_N23;5746;6825                          0    0             0   \n",
       "AM-294-K13_N4;7685;8899                           0   46             0   \n",
       "AM-294-L07_N4;47996;48265                         0    0             0   \n",
       "...                                             ...  ...           ...   \n",
       "AM-302-A23_N6;3581;4708                           0    0            46   \n",
       "AM-304-B23_N71;959;1417                           0    0             0   \n",
       "AM-307-C17_N9;25035;25322                         0    0             0   \n",
       "\n",
       "order                       Thermodesulfovibrionales  \n",
       "gene_cluster_rep                                      \n",
       "AM-294-J06_N23;5746;6825                           0  \n",
       "AM-294-K13_N4;7685;8899                        18569  \n",
       "AM-294-L07_N4;47996;48265                          0  \n",
       "...                                              ...  \n",
       "AM-302-A23_N6;3581;4708                            0  \n",
       "AM-304-B23_N71;959;1417                            1  \n",
       "AM-307-C17_N9;25035;25322                      24331  \n",
       "\n",
       "[14 rows x 9 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for file in read_files:\n",
    "    base=os.path.basename(file)\n",
    "    base=base[:-11]\n",
    "    Genus_name=base+'Genus_pivot.csv'\n",
    "    Family_name=base+'Family_pivot.csv'\n",
    "    Order_name=base+'Order_pivot.csv'\n",
    "    print(base)\n",
    "    \n",
    "    # I need to modify this so that I have one file that is a pivot of reads classified to the genus and one of reads classified to the fmaily level\n",
    "    \n",
    "    DF=pd.read_csv(file, index_col=0)\n",
    "    \n",
    "    if len (DF)==0:\n",
    "        continue\n",
    "    \n",
    "    else:\n",
    "        DF['taxonomic_lineage']=DF['taxonomic_lineage'].str[:-1]\n",
    "\n",
    "        DF[['domain', 'phyla', 'class', 'order', 'family', 'genus', 'species']]=DF.taxonomic_lineage.str.split(';', n=7, expand=True)\n",
    "        DF_Genus=DF.pivot_table(index='genus', columns='gene_cluster_rep', values='number_of_reads', aggfunc=np.sum, fill_value=0)\n",
    "        DF_Genus.to_csv(os.path.join(outdir, Genus_name))\n",
    "    \n",
    "        DF_Family=pd.pivot_table(DF, index='gene_cluster_rep', columns='family', values='number_of_reads', aggfunc=np.sum, fill_value=0)\n",
    "        DF_Family.to_csv(os.path.join(outdir, Family_name))\n",
    "        \n",
    "        DF_Order=pd.pivot_table(DF, index='gene_cluster_rep', columns='order', values='number_of_reads', aggfunc=np.sum, fill_value=0)\n",
    "        DF_Order.to_csv(os.path.join(outdir, Order_name))\n",
    "        \n",
    "DF_Order\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "6c398ef5-874e-4261-8a07-1a42a5d8c2e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>taxonomic_lineage</th>\n",
       "      <th>gene_cluster_rep</th>\n",
       "      <th>number_of_reads</th>\n",
       "      <th>prokka_EC_number</th>\n",
       "      <th>prokka_product</th>\n",
       "      <th>swissprot_gene</th>\n",
       "      <th>swissprot_EC_number</th>\n",
       "      <th>swissprot_eggNOG</th>\n",
       "      <th>swissprot_KO</th>\n",
       "      <th>swissprot_Pfam</th>\n",
       "      <th>swissprot_CAZy</th>\n",
       "      <th>swissprot_TIGRFAMs</th>\n",
       "      <th>domain</th>\n",
       "      <th>phyla</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>Archaea; Hadarchaeota; Hadarchaeia; Hadarchaea...</td>\n",
       "      <td>AM-294-L10_N20;891;1535</td>\n",
       "      <td>7457.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Hadarchaeota</td>\n",
       "      <td>Hadarchaeia</td>\n",
       "      <td>Hadarchaeales</td>\n",
       "      <td>WYZ-LMO6</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>Archaea; Thermoplasmatota; Thermoplasmata; Met...</td>\n",
       "      <td>AM-294-P13_N14;17572;18009</td>\n",
       "      <td>8336.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Archaea</td>\n",
       "      <td>Thermoplasmatota</td>\n",
       "      <td>Thermoplasmata</td>\n",
       "      <td>Methanomassiliicoccales</td>\n",
       "      <td>JACIVX01</td>\n",
       "      <td>NA</td>\n",
       "      <td>JACIVX01 sp011334735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>Bacteria; Acidobacteriota; Aminicenantia; Amin...</td>\n",
       "      <td>AM-294-P18_N23;17116;17334</td>\n",
       "      <td>32802.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Putative nitrogen fixation protein YutI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENOG41083MR,CCOG0694</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PF01106</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Acidobacteriota</td>\n",
       "      <td>Aminicenantia</td>\n",
       "      <td>Aminicenantales</td>\n",
       "      <td>DTGS01</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>13</td>\n",
       "      <td>Bacteria; Acidobacteriota; Aminicenantia; Amin...</td>\n",
       "      <td>AM-294-P18_N23;17116;17334</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Putative nitrogen fixation protein YutI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENOG41083MR,CCOG0694</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PF01106</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Acidobacteriota</td>\n",
       "      <td>Aminicenantia</td>\n",
       "      <td>Aminicenantales</td>\n",
       "      <td>Saccharicenantaceae</td>\n",
       "      <td>Saccharicenans</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>14</td>\n",
       "      <td>Bacteria; Firmicutes_B; GCA-003054495; Carboxy...</td>\n",
       "      <td>AM-294-J06_N23;5746;6825</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.3.5.5</td>\n",
       "      <td>Carbamoyl-phosphate synthase pyrimidine-specif...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PF00988,CPF00117</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TIGR01368</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes_B</td>\n",
       "      <td>GCA-003054495</td>\n",
       "      <td>Carboxydocellales</td>\n",
       "      <td>Carboxydocellaceae</td>\n",
       "      <td>Carboxydocella</td>\n",
       "      <td>Carboxydocella thermautotrophica</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>15</td>\n",
       "      <td>Bacteria; Nitrospirota; Thermodesulfovibrionia...</td>\n",
       "      <td>AM-296-B22_N14;2166;2447</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Nitrospirota</td>\n",
       "      <td>Thermodesulfovibrionia</td>\n",
       "      <td>Thermodesulfovibrionales</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Unnamed: 0                                  taxonomic_lineage  \\\n",
       "0            0  Archaea; Hadarchaeota; Hadarchaeia; Hadarchaea...   \n",
       "1            1  Archaea; Thermoplasmatota; Thermoplasmata; Met...   \n",
       "2            2  Bacteria; Acidobacteriota; Aminicenantia; Amin...   \n",
       "..         ...                                                ...   \n",
       "13          13  Bacteria; Acidobacteriota; Aminicenantia; Amin...   \n",
       "14          14  Bacteria; Firmicutes_B; GCA-003054495; Carboxy...   \n",
       "15          15  Bacteria; Nitrospirota; Thermodesulfovibrionia...   \n",
       "\n",
       "              gene_cluster_rep  number_of_reads prokka_EC_number  \\\n",
       "0      AM-294-L10_N20;891;1535           7457.0              NaN   \n",
       "1   AM-294-P13_N14;17572;18009           8336.0              NaN   \n",
       "2   AM-294-P18_N23;17116;17334          32802.0              NaN   \n",
       "..                         ...              ...              ...   \n",
       "13  AM-294-P18_N23;17116;17334              1.0              NaN   \n",
       "14    AM-294-J06_N23;5746;6825              1.0          6.3.5.5   \n",
       "15    AM-296-B22_N14;2166;2447              1.0              NaN   \n",
       "\n",
       "                                       prokka_product  swissprot_gene  \\\n",
       "0                                hypothetical protein             NaN   \n",
       "1                                hypothetical protein             NaN   \n",
       "2             Putative nitrogen fixation protein YutI             NaN   \n",
       "..                                                ...             ...   \n",
       "13            Putative nitrogen fixation protein YutI             NaN   \n",
       "14  Carbamoyl-phosphate synthase pyrimidine-specif...             NaN   \n",
       "15                               hypothetical protein             NaN   \n",
       "\n",
       "    swissprot_EC_number      swissprot_eggNOG swissprot_KO    swissprot_Pfam  \\\n",
       "0                   NaN                   NaN          NaN               NaN   \n",
       "1                   NaN                   NaN          NaN               NaN   \n",
       "2                   NaN  ENOG41083MR,CCOG0694          NaN           PF01106   \n",
       "..                  ...                   ...          ...               ...   \n",
       "13                  NaN  ENOG41083MR,CCOG0694          NaN           PF01106   \n",
       "14                  NaN                   NaN          NaN  PF00988,CPF00117   \n",
       "15                  NaN                   NaN          NaN               NaN   \n",
       "\n",
       "    swissprot_CAZy swissprot_TIGRFAMs    domain              phyla  \\\n",
       "0              NaN                NaN   Archaea       Hadarchaeota   \n",
       "1              NaN                NaN   Archaea   Thermoplasmatota   \n",
       "2              NaN                NaN  Bacteria    Acidobacteriota   \n",
       "..             ...                ...       ...                ...   \n",
       "13             NaN                NaN  Bacteria    Acidobacteriota   \n",
       "14             NaN          TIGR01368  Bacteria       Firmicutes_B   \n",
       "15             NaN                NaN  Bacteria       Nitrospirota   \n",
       "\n",
       "                      class                      order                family  \\\n",
       "0               Hadarchaeia              Hadarchaeales              WYZ-LMO6   \n",
       "1            Thermoplasmata    Methanomassiliicoccales              JACIVX01   \n",
       "2             Aminicenantia            Aminicenantales                DTGS01   \n",
       "..                      ...                        ...                   ...   \n",
       "13            Aminicenantia            Aminicenantales   Saccharicenantaceae   \n",
       "14            GCA-003054495          Carboxydocellales    Carboxydocellaceae   \n",
       "15   Thermodesulfovibrionia   Thermodesulfovibrionales                    NA   \n",
       "\n",
       "              genus                            species  \n",
       "0                NA                                 NA  \n",
       "1                NA               JACIVX01 sp011334735  \n",
       "2                NA                                 NA  \n",
       "..              ...                                ...  \n",
       "13   Saccharicenans                                 NA  \n",
       "14   Carboxydocella   Carboxydocella thermautotrophica  \n",
       "15               NA                                 NA  \n",
       "\n",
       "[16 rows x 20 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test=pd.read_csv('/mnt/scgc/simon/microg2p/Melody/BLM1/Analyses/RNA_GORG_recruit/Summary_files/cDNA1-A01_joined_annotated_RNA_GORG_results.csv')\n",
    "test['taxonomic_lineage']=test['taxonomic_lineage'].str[:-1]\n",
    "test[['domain', 'phyla', 'class', 'order', 'family', 'genus', 'species']]=test.taxonomic_lineage.str.split(';', expand=True)\n",
    "\n",
    "test\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7292d2a-15a2-4b72-90fb-6d4dceebb3bc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d123015f-cc52-44bf-9bfc-5ba668db8f5e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "befd326e-a0ec-462b-ba2b-969ff4676513",
   "metadata": {},
   "outputs": [],
   "source": [
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aaec8b93-1e31-43e6-98bd-83dc0325125d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "jacob_env (julia-clone)",
   "language": "python",
   "name": "jacob_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
