{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "cc4598b9-495e-4410-97d4-f65dfa20949f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Folder already exists\n",
      "Folder already exists\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'3.1.0'"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import glob\n",
    "from collections import defaultdict\n",
    "import pandas as pd\n",
    "import os.path as op\n",
    "import math\n",
    "import os \n",
    "import sys\n",
    "import numpy as np\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from collections import Counter\n",
    "\n",
    "import sqlite3\n",
    "import sqlalchemy\n",
    "\n",
    "def safe_makedir(dname):\n",
    "    \"\"\"\n",
    "    Make a directory if it doesn't exist, handling concurrent race conditions.\n",
    "    \"\"\"\n",
    "    if not dname:\n",
    "        return dname\n",
    "    num_tries = 0\n",
    "    max_tries = 5\n",
    "    while not os.path.exists(dname):\n",
    "        try:\n",
    "            os.makedirs(dname)\n",
    "        except OSError:\n",
    "            if num_tries > max_tries:\n",
    "                raise\n",
    "            num_tries += 1\n",
    "            time.sleep(2)\n",
    "    return dname\n",
    "\n",
    "# import my custom functions\n",
    "sys.path.insert(0, '/home/jmunson-mcgee/')\n",
    "from JMM_functions import *\n",
    "\n",
    "pd.set_option(\"display.max_rows\", 6)\n",
    "\n",
    "outdir='/mnt/scgc/simon/microg2p/Melody/BLM1/Analyses/RNA_GORG_recruit/Summary_files/'\n",
    "safe_make_dir(outdir)\n",
    "os.chdir(outdir)\n",
    "\n",
    "images='/mnt/scgc/simon/microg2p/Melody/BLM1/images/'\n",
    "safe_make_dir(images)\n",
    "    \n",
    "matplotlib.__version__\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "7be8d4a8-8090-4cc5-8147-779ab011c2b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "libraries={'cDNA1-A01':'No_amendment_transcripts', 'cDNA2-A01':'Acetate_transcripts', 'cDNA3-A01':'CO_transcripts', \n",
    "         'cDNA4-A01':'H2_transcripts', 'cDNA7-A01':'Manifold_transcripts', 'cDNA6-A01':'cDNA6_control', 'cDNA9-A01':'cDNA9_control'}\n",
    "\n",
    "in_files=glob.glob('/mnt/scgc/simon/microg2p/Melody/BLM1/Analyses/RNA_GORG_recruit/Summary_files/*_joined_annotated_RNA_GORG_results.csv')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "aa24c3f1-08c0-43cf-a450-7d85765d40dc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cDNA4-A01\n",
      "the number of unique gene clusters is 11\n",
      "11\n",
      "cDNA2-A01\n",
      "the number of unique gene clusters is 12\n",
      "14\n",
      "cDNA3-A01\n",
      "the number of unique gene clusters is 2\n",
      "2\n",
      "cDNA7-A01\n",
      "the number of unique gene clusters is 140\n",
      "153\n",
      "cDNA1-A01\n",
      "the number of unique gene clusters is 14\n",
      "16\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/scgc_nfs/opt/common/anaconda3a/lib/python3.7/site-packages/ipykernel_launcher.py:16: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n",
      "  app.launch_new_instance()\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>amendment</th>\n",
       "      <th>gene_cluster_rep</th>\n",
       "      <th>number_of_reads</th>\n",
       "      <th>prokka_EC_number</th>\n",
       "      <th>prokka_product</th>\n",
       "      <th>swissprot_CAZy</th>\n",
       "      <th>swissprot_EC_number</th>\n",
       "      <th>swissprot_KO</th>\n",
       "      <th>swissprot_Pfam</th>\n",
       "      <th>swissprot_TIGRFAMs</th>\n",
       "      <th>swissprot_eggNOG</th>\n",
       "      <th>swissprot_gene</th>\n",
       "      <th>taxonomic_lineage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>H2_transcripts</td>\n",
       "      <td>AM-294-G05_N10;1964;2641</td>\n",
       "      <td>28722.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Archaea; Hadarchaeota; Hadarchaeia; Hadarchaea...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>H2_transcripts</td>\n",
       "      <td>AM-294-I22_N59;2566;2784</td>\n",
       "      <td>66.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Archaea; Hadarchaeota; Hadarchaeia; Hadarchaea...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>H2_transcripts</td>\n",
       "      <td>AM-294-G10_N3;30105;32429</td>\n",
       "      <td>18923.0</td>\n",
       "      <td>1.17.1.4</td>\n",
       "      <td>Probable xanthine dehydrogenase subunit D (XDH...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>K00087</td>\n",
       "      <td>PF01315,CPF02738</td>\n",
       "      <td>TIGR03196</td>\n",
       "      <td>ENOG4105CE2,CCOG1529</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria; Bipolaricaulota; Bipolaricaulia; Bip...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>NaN</td>\n",
       "      <td>No_amendment_transcripts</td>\n",
       "      <td>AM-294-P18_N23;17116;17334</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Putative nitrogen fixation protein YutI</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PF01106</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENOG41083MR,CCOG0694</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria; Acidobacteriota; Aminicenantia; Amin...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>NaN</td>\n",
       "      <td>No_amendment_transcripts</td>\n",
       "      <td>AM-294-J06_N23;5746;6825</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.3.5.5</td>\n",
       "      <td>Carbamoyl-phosphate synthase pyrimidine-specif...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PF00988,CPF00117</td>\n",
       "      <td>TIGR01368</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria; Firmicutes_B; GCA-003054495; Carboxy...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>NaN</td>\n",
       "      <td>No_amendment_transcripts</td>\n",
       "      <td>AM-296-B22_N14;2166;2447</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria; Nitrospirota; Thermodesulfovibrionia...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>196 rows Ã— 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      0                 amendment            gene_cluster_rep  \\\n",
       "0   NaN            H2_transcripts    AM-294-G05_N10;1964;2641   \n",
       "1   NaN            H2_transcripts    AM-294-I22_N59;2566;2784   \n",
       "2   NaN            H2_transcripts   AM-294-G10_N3;30105;32429   \n",
       "..   ..                       ...                         ...   \n",
       "193 NaN  No_amendment_transcripts  AM-294-P18_N23;17116;17334   \n",
       "194 NaN  No_amendment_transcripts    AM-294-J06_N23;5746;6825   \n",
       "195 NaN  No_amendment_transcripts    AM-296-B22_N14;2166;2447   \n",
       "\n",
       "     number_of_reads prokka_EC_number  \\\n",
       "0            28722.0              NaN   \n",
       "1               66.0              NaN   \n",
       "2            18923.0         1.17.1.4   \n",
       "..               ...              ...   \n",
       "193              1.0              NaN   \n",
       "194              1.0          6.3.5.5   \n",
       "195              1.0              NaN   \n",
       "\n",
       "                                        prokka_product swissprot_CAZy  \\\n",
       "0                                 hypothetical protein            NaN   \n",
       "1                                 hypothetical protein            NaN   \n",
       "2    Probable xanthine dehydrogenase subunit D (XDH...            NaN   \n",
       "..                                                 ...            ...   \n",
       "193            Putative nitrogen fixation protein YutI            NaN   \n",
       "194  Carbamoyl-phosphate synthase pyrimidine-specif...            NaN   \n",
       "195                               hypothetical protein            NaN   \n",
       "\n",
       "     swissprot_EC_number swissprot_KO    swissprot_Pfam swissprot_TIGRFAMs  \\\n",
       "0                    NaN          NaN               NaN                NaN   \n",
       "1                    NaN          NaN               NaN                NaN   \n",
       "2                    NaN       K00087  PF01315,CPF02738          TIGR03196   \n",
       "..                   ...          ...               ...                ...   \n",
       "193                  NaN          NaN           PF01106                NaN   \n",
       "194                  NaN          NaN  PF00988,CPF00117          TIGR01368   \n",
       "195                  NaN          NaN               NaN                NaN   \n",
       "\n",
       "         swissprot_eggNOG  swissprot_gene  \\\n",
       "0                     NaN             NaN   \n",
       "1                     NaN             NaN   \n",
       "2    ENOG4105CE2,CCOG1529             NaN   \n",
       "..                    ...             ...   \n",
       "193  ENOG41083MR,CCOG0694             NaN   \n",
       "194                   NaN             NaN   \n",
       "195                   NaN             NaN   \n",
       "\n",
       "                                     taxonomic_lineage  \n",
       "0    Archaea; Hadarchaeota; Hadarchaeia; Hadarchaea...  \n",
       "1    Archaea; Hadarchaeota; Hadarchaeia; Hadarchaea...  \n",
       "2    Bacteria; Bipolaricaulota; Bipolaricaulia; Bip...  \n",
       "..                                                 ...  \n",
       "193  Bacteria; Acidobacteriota; Aminicenantia; Amin...  \n",
       "194  Bacteria; Firmicutes_B; GCA-003054495; Carboxy...  \n",
       "195  Bacteria; Nitrospirota; Thermodesulfovibrionia...  \n",
       "\n",
       "[196 rows x 14 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_master=pd.DataFrame()\n",
    "\n",
    "for file in in_files:\n",
    "    base=os.path.basename(file)\n",
    "    sample=base[:9]\n",
    "    amend=libraries[sample]\n",
    "    print(sample)\n",
    "    \n",
    "    DF_tmp=pd.read_csv(file, index_col=0)\n",
    "    DF_tmp['amendment']=amend\n",
    "    n = DF_tmp.gene_cluster_rep.nunique()\n",
    "    print('the number of unique gene clusters is', n)\n",
    "    \n",
    "    print(len(DF_tmp))\n",
    "    #if len(DF_tmp)>1\n",
    "    DF_master=pd.concat([DF_master, DF_tmp], ignore_index=True)\n",
    "    \n",
    "DF_master.to_csv(os.path.join(outdir, 'BLM1_combined_GORG_recruitments.csv'))\n",
    "\n",
    "DF_master"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "9e941db2-31c7-4e50-aefe-47b9ce4bd96e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the number of unique gene clusters is 173\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "191"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "n = DF_master.gene_cluster_rep.nunique()\n",
    "print('the number of unique gene clusters is', n)\n",
    "\n",
    "DF_master.groupby(['gene_cluster_rep', 'taxonomic_lineage']).ngroups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "c9919f75-ec67-47a4-b585-6a86c5a1e7e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>amendment</th>\n",
       "      <th>gene_cluster_rep</th>\n",
       "      <th>number_of_reads</th>\n",
       "      <th>prokka_EC_number</th>\n",
       "      <th>prokka_product</th>\n",
       "      <th>swissprot_CAZy</th>\n",
       "      <th>swissprot_EC_number</th>\n",
       "      <th>swissprot_KO</th>\n",
       "      <th>swissprot_Pfam</th>\n",
       "      <th>swissprot_TIGRFAMs</th>\n",
       "      <th>swissprot_eggNOG</th>\n",
       "      <th>swissprot_gene</th>\n",
       "      <th>taxonomic_lineage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Manifold_transcripts</td>\n",
       "      <td>AM-294-N05_N9;10214;11533</td>\n",
       "      <td>22600.0</td>\n",
       "      <td>3.1.-.-</td>\n",
       "      <td>CRISPR-associated endonuclease Cas9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria; Firmicutes_B; Desulfotomaculia; Ammo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Manifold_transcripts</td>\n",
       "      <td>AM-294-N05_N9;10214;11533</td>\n",
       "      <td>76.0</td>\n",
       "      <td>3.1.-.-</td>\n",
       "      <td>CRISPR-associated endonuclease Cas9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria; CSP1-3; NA; NA; SpSt-165; NA; SpSt-1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Manifold_transcripts</td>\n",
       "      <td>AM-294-N20_N7;20661;21965</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.1.-.-</td>\n",
       "      <td>CRISPR-associated endonuclease Cas9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacteria; Firmicutes_B; Desulfotomaculia; Ammo...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0             amendment           gene_cluster_rep  number_of_reads  \\\n",
       "70  NaN  Manifold_transcripts  AM-294-N05_N9;10214;11533          22600.0   \n",
       "97  NaN  Manifold_transcripts  AM-294-N05_N9;10214;11533             76.0   \n",
       "149 NaN  Manifold_transcripts  AM-294-N20_N7;20661;21965              4.0   \n",
       "\n",
       "    prokka_EC_number                       prokka_product swissprot_CAZy  \\\n",
       "70           3.1.-.-  CRISPR-associated endonuclease Cas9            NaN   \n",
       "97           3.1.-.-  CRISPR-associated endonuclease Cas9            NaN   \n",
       "149          3.1.-.-  CRISPR-associated endonuclease Cas9            NaN   \n",
       "\n",
       "     swissprot_EC_number swissprot_KO swissprot_Pfam swissprot_TIGRFAMs  \\\n",
       "70                   NaN          NaN            NaN                NaN   \n",
       "97                   NaN          NaN            NaN                NaN   \n",
       "149                  NaN          NaN            NaN                NaN   \n",
       "\n",
       "    swissprot_eggNOG  swissprot_gene  \\\n",
       "70               NaN             NaN   \n",
       "97               NaN             NaN   \n",
       "149              NaN             NaN   \n",
       "\n",
       "                                     taxonomic_lineage  \n",
       "70   Bacteria; Firmicutes_B; Desulfotomaculia; Ammo...  \n",
       "97   Bacteria; CSP1-3; NA; NA; SpSt-165; NA; SpSt-1...  \n",
       "149  Bacteria; Firmicutes_B; Desulfotomaculia; Ammo...  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_Cas9=DF_master.copy()\n",
    "DF_Cas9['prokka_product']=DF_Cas9['prokka_product'].fillna('hypothetical protein')\n",
    "DF_Cas9=DF_Cas9[DF_Cas9['prokka_product'].str.contains('Cas9')]\n",
    "DF_Cas9.to_csv(os.path.join(outdir, 'BLM1_cas9_transcripts'))\n",
    "DF_Cas9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e84bc697-d533-4766-b1e8-2ed88ead2ffb",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Jacob",
   "language": "python",
   "name": "jacob"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
